{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "819bd65b-f5b5-4644-a058-0a8c843d4fe8",
   "metadata": {},
   "source": [
    "# Observed Taylor law: variance vs mean \n",
    "**in protein coding gene length distributions for the different species**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8046a185-31f3-48b1-83f7-947911a96d15",
   "metadata": {},
   "source": [
    "## Import python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "51b0a90c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d509543a-9bbf-4428-bb98-a287201ef79d",
   "metadata": {},
   "source": [
    "## Retrieving the statistical descriptions of the gene length's distributions for the different genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e96ab72e-e702-4020-814a-625c86dd44ee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The statistical descriptions of the protein coding gene distributions for the different species is in:\n",
      " /home/emuro/stat_protCodGenes_with_ncbiGenomeData.tsv \n",
      "\n",
      "(33627, 13)\n",
      "division_7\n",
      "bacteria       31943\n",
      "fungi           1014\n",
      "protists         237\n",
      "vertebrates      222\n",
      "metazoa          115\n",
      "plants            96\n",
      "Name: count, dtype: int64\n",
      "group\n",
      "Proteobacteria                     14417\n",
      "Terrabacteria group                10997\n",
      "FCB group                           2601\n",
      "NaN                                 1605\n",
      "Fungi                                980\n",
      "Euryarchaeota                        686\n",
      "PVC group                            308\n",
      "TACK group                           291\n",
      "Animals                              274\n",
      "Spirochaetes                         221\n",
      "Protists                             211\n",
      "Acidobacteria                        190\n",
      "DPANN group                          122\n",
      "Bacteria incertae sedis              118\n",
      "Fusobacteria                         110\n",
      "Plants                                70\n",
      "Thermotogae                           66\n",
      "unclassified Bacteria                 63\n",
      "Candidatus Thermoplasmatota           63\n",
      "Nitrospirae                           55\n",
      "Synergistetes                         25\n",
      "Asgard group                          24\n",
      "Aquificae                             24\n",
      "unclassified Archaea                  17\n",
      "Elusimicrobia                         17\n",
      "Nitrospinae/Tectomicrobia group       15\n",
      "Other                                 15\n",
      "Calditrichaeota                        9\n",
      "Atribacterota                          9\n",
      "Deferribacteres                        8\n",
      "Thermodesulfobacteria                  7\n",
      "environmental samples                  4\n",
      "Dictyoglomi                            3\n",
      "Chrysiogenetes                         1\n",
      "Coprothermobacterota                   1\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>division_7</th>\n",
       "      <th>division_8</th>\n",
       "      <th>species</th>\n",
       "      <th>assembly</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>ensembl_assembly_accession</th>\n",
       "      <th>group</th>\n",
       "      <th>subgroup</th>\n",
       "      <th>gc_percent</th>\n",
       "      <th>size_Mbp</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>chromosomes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>archaea</td>\n",
       "      <td>methanobacterium_bryantii_gca_002287175</td>\n",
       "      <td>ASM228717v1</td>\n",
       "      <td>2161</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "      <td>GCA_002287175.1</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomada group</td>\n",
       "      <td>33.2</td>\n",
       "      <td>3.46637</td>\n",
       "      <td>Contig</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>archaea</td>\n",
       "      <td>methanobacterium_formicicum_gca_000762265</td>\n",
       "      <td>ASM76226v1</td>\n",
       "      <td>2162</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "      <td>GCA_000762265.1</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomada group</td>\n",
       "      <td>41.3</td>\n",
       "      <td>2.44999</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>chromosome:NZ_CP006933.1/CP006933.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  division_7 division_8                                    species  \\\n",
       "0   bacteria    archaea    methanobacterium_bryantii_gca_002287175   \n",
       "1   bacteria    archaea  methanobacterium_formicicum_gca_000762265   \n",
       "\n",
       "      assembly  taxonomy_id  \\\n",
       "0  ASM228717v1         2161   \n",
       "1   ASM76226v1         2162   \n",
       "\n",
       "                                             Lineage  \\\n",
       "0  Archaea; Euryarchaeota; Methanomada group; Met...   \n",
       "1  Archaea; Euryarchaeota; Methanomada group; Met...   \n",
       "\n",
       "  ensembl_assembly_accession          group           subgroup gc_percent  \\\n",
       "0            GCA_002287175.1  Euryarchaeota  Methanomada group       33.2   \n",
       "1            GCA_000762265.1  Euryarchaeota  Methanomada group       41.3   \n",
       "\n",
       "   size_Mbp  assembly_status                          chromosomes  \n",
       "0   3.46637           Contig                                    -  \n",
       "1   2.44999  Complete Genome  chromosome:NZ_CP006933.1/CP006933.1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# statistics on length distribution for different species\n",
    "stat_file = \"/home/emuro/stat_protCodGenes_with_ncbiGenomeData.tsv\"\n",
    "print(\"The statistical descriptions of the protein coding gene distributions for the different species is in:\\n\", stat_file, \"\\n\")\n",
    "\n",
    "# retrieve data and diminish the number of columns\n",
    "stat_df = pd.read_csv(stat_file, low_memory=False, sep=\"\\t\")\n",
    "#stat_df = stat_df[[\"division_7\", \"division_8\", \"species\", \"taxonomy_id\", \"Lineage\", \"ensembl_assembly_accession\", \"group\", \"subgroup\", \"assembly_status\"]] \n",
    "\n",
    "##stat_df = stat_df[[\"division_7\", \"division_8\", \"species\", \"taxonomy_id\", \"Lineage\", \"group\", \"subgroup\"]] \n",
    "\n",
    "# visualize the data frame\n",
    "if 1:\n",
    "    print(stat_df.shape)\n",
    "    print(stat_df.value_counts('division_7', dropna=False))\n",
    "    print(stat_df.value_counts('group',      dropna=False))\n",
    "    pd.set_option('display.max_columns',     stat_df.shape[1])\n",
    "    display(stat_df.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7a9f831b-bf62-4c33-9411-fc4c9438bc17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(33627, 13)\n",
      "division_7\n",
      "bacteria       31943\n",
      "fungi           1014\n",
      "protists         237\n",
      "vertebrates      222\n",
      "metazoa          115\n",
      "plants            96\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>division_7</th>\n",
       "      <th>division_8</th>\n",
       "      <th>species</th>\n",
       "      <th>assembly</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>ensembl_assembly_accession</th>\n",
       "      <th>group</th>\n",
       "      <th>subgroup</th>\n",
       "      <th>gc_percent</th>\n",
       "      <th>size_Mbp</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>chromosomes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>archaea</td>\n",
       "      <td>methanobacterium_bryantii_gca_002287175</td>\n",
       "      <td>ASM228717v1</td>\n",
       "      <td>2161</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "      <td>GCA_002287175.1</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomada group</td>\n",
       "      <td>33.2</td>\n",
       "      <td>3.46637</td>\n",
       "      <td>Contig</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>archaea</td>\n",
       "      <td>methanobacterium_formicicum_gca_000762265</td>\n",
       "      <td>ASM76226v1</td>\n",
       "      <td>2162</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "      <td>GCA_000762265.1</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Methanomada group</td>\n",
       "      <td>41.3</td>\n",
       "      <td>2.44999</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>chromosome:NZ_CP006933.1/CP006933.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  division_7 division_8                                    species  \\\n",
       "0   bacteria    archaea    methanobacterium_bryantii_gca_002287175   \n",
       "1   bacteria    archaea  methanobacterium_formicicum_gca_000762265   \n",
       "\n",
       "      assembly  taxonomy_id  \\\n",
       "0  ASM228717v1         2161   \n",
       "1   ASM76226v1         2162   \n",
       "\n",
       "                                             Lineage  \\\n",
       "0  Archaea; Euryarchaeota; Methanomada group; Met...   \n",
       "1  Archaea; Euryarchaeota; Methanomada group; Met...   \n",
       "\n",
       "  ensembl_assembly_accession          group           subgroup gc_percent  \\\n",
       "0            GCA_002287175.1  Euryarchaeota  Methanomada group       33.2   \n",
       "1            GCA_000762265.1  Euryarchaeota  Methanomada group       41.3   \n",
       "\n",
       "   size_Mbp  assembly_status                          chromosomes  \n",
       "0   3.46637           Contig                                    -  \n",
       "1   2.44999  Complete Genome  chromosome:NZ_CP006933.1/CP006933.1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##stat_df[\"Fer_group\"] = np.nan\n",
    "# visualize the data frame\n",
    "if 1:\n",
    "    print(stat_df.shape)\n",
    "    print(stat_df.value_counts('division_7', dropna=False))\n",
    "    pd.set_option('display.max_columns', stat_df.shape[1])\n",
    "    display(stat_df.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "082cd0f5-3d69-40df-87b2-7cc0e55ae334",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "56367958-49f7-4731-8b43-8f28a6513ce1",
   "metadata": {},
   "source": [
    "## Retrieving the data from Fer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f53939ad-2416-486e-8db2-8dcccf3e5d54",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The statistical descriptions of the protein coding gene distributions for the different species is in:\n",
      " /home/emuro/git/github/EM_geneLength_nature/main_tables/extra_tables/genes.xlsx \n",
      "\n",
      "(222, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>genes_mean</th>\n",
       "      <th>genes_var</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acanthochromis_polyacanthus</td>\n",
       "      <td>17577.741964</td>\n",
       "      <td>7.636914e+08</td>\n",
       "      <td>fishes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>amphilophus_citrinellus</td>\n",
       "      <td>13673.609124</td>\n",
       "      <td>5.864511e+08</td>\n",
       "      <td>fishes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       species    genes_mean     genes_var   group\n",
       "0  acanthochromis_polyacanthus  17577.741964  7.636914e+08  fishes\n",
       "1      amphilophus_citrinellus  13673.609124  5.864511e+08  fishes"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# statistics on length distribution for different species\n",
    "fer_file = \"/home/emuro/git/github/EM_geneLength_nature/main_tables/extra_tables/genes.xlsx\"\n",
    "print(\"The statistical descriptions of the protein coding gene distributions for the different species is in:\\n\", fer_file, \"\\n\")\n",
    "\n",
    "# retrieve data and diminish the number of columns\n",
    "fer_df = pd.read_excel(fer_file, sheet_name=\"Vertebrata\")\n",
    "#fer_df = fer_df[[\"division_7\", \"division_8\", \"species\", \"taxonomy_id\", \"Lineage\", \"ensembl_assembly_accession\", \"group\", \"subgroup\", \"assembly_status\"]] \n",
    "\n",
    "# visualize the data frame\n",
    "if 1:\n",
    "    print(fer_df.shape)\n",
    "    #print(fer_df.value_counts('division_7', dropna=False))\n",
    "    pd.set_option('display.max_columns', stat_df.shape[1])\n",
    "    display(fer_df.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f55f17e-5351-45aa-a294-93e150f29b01",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4810cb0d-9954-4541-810a-bee1db04fb26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(33627, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>division_7</th>\n",
       "      <th>division_8</th>\n",
       "      <th>species</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>group</th>\n",
       "      <th>subgroup</th>\n",
       "      <th>group_Fer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33238</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>petromyzon_marinus</td>\n",
       "      <td>7757</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fishes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33239</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>eptatretus_burgeri</td>\n",
       "      <td>7764</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Fishes</td>\n",
       "      <td>fishes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        division_7   division_8             species  taxonomy_id  \\\n",
       "33238  vertebrates  vertebrates  petromyzon_marinus         7757   \n",
       "33239  vertebrates  vertebrates  eptatretus_burgeri         7764   \n",
       "\n",
       "                                                 Lineage    group subgroup  \\\n",
       "33238  Eukaryota; Metazoa; Chordata; Craniata; Verteb...      NaN      NaN   \n",
       "33239  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  Animals   Fishes   \n",
       "\n",
       "      group_Fer  \n",
       "33238    fishes  \n",
       "33239    fishes  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "merged_df = pd.merge(stat_df, fer_df, on=\"species\", how=\"left\")\n",
    "merged_df = merged_df.rename(columns={'group_x': 'group', 'group_y': 'group_Fer'})\n",
    "df_cp = merged_df[[\"division_7\", \"division_8\", \"species\", \"taxonomy_id\", \"Lineage\", \"group\", \"subgroup\", \"group_Fer\"]]\n",
    "\n",
    "df_cp = df_cp[df_cp.division_7==\"vertebrates\"]\n",
    "# visualize the data frame\n",
    "if 1:\n",
    "    print(merged_df.shape)\n",
    "    #print(fer_df.value_counts('division_7', dropna=False))\n",
    "    pd.set_option('display.max_columns', merged_df.shape[1])\n",
    "    #pd.set_option('display.max_rows', 222)\n",
    "    #pd.set_option('display.max_colwidth', None)\n",
    "    display(df_cp[df_cp.division_7==\"vertebrates\"].head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82e29dc1-e16e-4b32-aab4-ef2eee0a4917",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aa8f75b4-cdea-4d26-83e4-e8c017ce4dd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the file if desired\n",
    "if 0:\n",
    "    merged_df.to_csv(\"/home/emuro/stat_protCodGenes_with_ncbiGenomeData_FerGroup.tsv\", sep=\"\\t\", index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcd14357-9cfa-41ca-bb6e-097985d55c6c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
