{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4e158002-7422-422c-89ca-fbaacb381b27",
   "metadata": {},
   "source": [
    "# Distribution of the lengths of the proteins\n",
    "**for the different species (reference proteomes)**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8046a185-31f3-48b1-83f7-947911a96d15",
   "metadata": {},
   "source": [
    "## Import python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "51b0a90c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "from scipy import stats \n",
    "import os\n",
    "import sys\n",
    "sys.path.append('../../lib/')\n",
    "import constants as c\n",
    "\n",
    "\n",
    "BOOL_SAVE_FILE = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7b6d3b5",
   "metadata": {},
   "source": [
    "## The protein length distributions for the different species"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d509543a-9bbf-4428-bb98-a287201ef79d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Retrieving the statistical descriptions of the length's distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "86b76b58-6c34-4b8a-82b3-f117e835fad0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The statistical descriptions of the protein distributions for the different species is in:\n",
      " /home/emuro/git/github/EM_geneLength_nature/working_on/preprocessing_data/../../lib/../main_tables/stat_merged.tsv \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes_species</th>\n",
       "      <th>genes_assembly</th>\n",
       "      <th>genes_trunk_genes_path</th>\n",
       "      <th>genes_genes_file</th>\n",
       "      <th>genes_count</th>\n",
       "      <th>genes_mean</th>\n",
       "      <th>genes_std</th>\n",
       "      <th>genes_var</th>\n",
       "      <th>genes_min</th>\n",
       "      <th>genes_25perc</th>\n",
       "      <th>genes_50perc</th>\n",
       "      <th>genes_75perc</th>\n",
       "      <th>genes_max</th>\n",
       "      <th>genes_log10_mean</th>\n",
       "      <th>genes_log10_std</th>\n",
       "      <th>genes_log10_var</th>\n",
       "      <th>genes_log10_min</th>\n",
       "      <th>genes_log10_25perc</th>\n",
       "      <th>genes_log10_50perc</th>\n",
       "      <th>genes_log10_75perc</th>\n",
       "      <th>genes_log10_max</th>\n",
       "      <th>genes_log_mean</th>\n",
       "      <th>genes_log_std</th>\n",
       "      <th>genes_log_var</th>\n",
       "      <th>genes_log_min</th>\n",
       "      <th>genes_log_25perc</th>\n",
       "      <th>genes_log_50perc</th>\n",
       "      <th>genes_log_75perc</th>\n",
       "      <th>genes_log_max</th>\n",
       "      <th>genes_division</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>prots_species</th>\n",
       "      <th>prots_proteome_id</th>\n",
       "      <th>prots_superregnum</th>\n",
       "      <th>prots_num_prot_cod_genes</th>\n",
       "      <th>prots_uniprot_fasta_file</th>\n",
       "      <th>prots_count</th>\n",
       "      <th>prots_mean</th>\n",
       "      <th>prots_std</th>\n",
       "      <th>prots_var</th>\n",
       "      <th>prots_min</th>\n",
       "      <th>prots_25perc</th>\n",
       "      <th>prots_50perc</th>\n",
       "      <th>prots_75perc</th>\n",
       "      <th>prots_max</th>\n",
       "      <th>prots_log10_mean</th>\n",
       "      <th>prots_log10_std</th>\n",
       "      <th>prots_log10_var</th>\n",
       "      <th>prots_log10_min</th>\n",
       "      <th>prots_log10_25perc</th>\n",
       "      <th>prots_log10_50perc</th>\n",
       "      <th>prots_log10_75perc</th>\n",
       "      <th>prots_log10_max</th>\n",
       "      <th>prots_log_mean</th>\n",
       "      <th>prots_log_std</th>\n",
       "      <th>prots_log_var</th>\n",
       "      <th>prots_log_min</th>\n",
       "      <th>prots_log_25perc</th>\n",
       "      <th>prots_log_50perc</th>\n",
       "      <th>prots_log_75perc</th>\n",
       "      <th>prots_log_max</th>\n",
       "      <th>merged_division_superregnum</th>\n",
       "      <th>ratio_mean_prots_genes</th>\n",
       "      <th>diff_prots_genes</th>\n",
       "      <th>abs_diff</th>\n",
       "      <th>ratio_prots_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>archangium_gephyra_gca_001027285</td>\n",
       "      <td>ASM102728v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.archangium_gephyra_gca_00...</td>\n",
       "      <td>10119.0</td>\n",
       "      <td>1103.871035</td>\n",
       "      <td>1127.066255</td>\n",
       "      <td>1.270278e+06</td>\n",
       "      <td>113.0</td>\n",
       "      <td>524.0</td>\n",
       "      <td>887.0</td>\n",
       "      <td>1388.0</td>\n",
       "      <td>36653.0</td>\n",
       "      <td>2.919254</td>\n",
       "      <td>0.333869</td>\n",
       "      <td>0.111468</td>\n",
       "      <td>2.053078</td>\n",
       "      <td>2.719331</td>\n",
       "      <td>2.947924</td>\n",
       "      <td>3.142389</td>\n",
       "      <td>4.56411</td>\n",
       "      <td>6.721831</td>\n",
       "      <td>0.768762</td>\n",
       "      <td>0.590995</td>\n",
       "      <td>4.727388</td>\n",
       "      <td>6.261492</td>\n",
       "      <td>6.787845</td>\n",
       "      <td>7.235619</td>\n",
       "      <td>10.509251</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>48</td>\n",
       "      <td>Archangium gephyra</td>\n",
       "      <td>UP000035579</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>10110</td>\n",
       "      <td>/ftp.uniprot.org/pub/databases/uniprot/current...</td>\n",
       "      <td>10110.0</td>\n",
       "      <td>367.499703</td>\n",
       "      <td>375.772244</td>\n",
       "      <td>141204.779058</td>\n",
       "      <td>37.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>295.0</td>\n",
       "      <td>462.0</td>\n",
       "      <td>12217.0</td>\n",
       "      <td>2.441094</td>\n",
       "      <td>0.334836</td>\n",
       "      <td>0.112115</td>\n",
       "      <td>1.568202</td>\n",
       "      <td>2.240549</td>\n",
       "      <td>2.469822</td>\n",
       "      <td>2.664642</td>\n",
       "      <td>4.086965</td>\n",
       "      <td>5.620827</td>\n",
       "      <td>0.770987</td>\n",
       "      <td>0.594422</td>\n",
       "      <td>3.610918</td>\n",
       "      <td>5.159055</td>\n",
       "      <td>5.686975</td>\n",
       "      <td>6.135565</td>\n",
       "      <td>9.410584</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>0.998757</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.999111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chondromyces_crocatus_gca_001189295</td>\n",
       "      <td>ASM118929v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.chondromyces_crocatus_gca...</td>\n",
       "      <td>8339.0</td>\n",
       "      <td>1211.514450</td>\n",
       "      <td>1051.408605</td>\n",
       "      <td>1.105460e+06</td>\n",
       "      <td>89.0</td>\n",
       "      <td>629.0</td>\n",
       "      <td>977.0</td>\n",
       "      <td>1460.0</td>\n",
       "      <td>24476.0</td>\n",
       "      <td>2.978963</td>\n",
       "      <td>0.300250</td>\n",
       "      <td>0.090150</td>\n",
       "      <td>1.949390</td>\n",
       "      <td>2.798651</td>\n",
       "      <td>2.989895</td>\n",
       "      <td>3.164353</td>\n",
       "      <td>4.38874</td>\n",
       "      <td>6.859315</td>\n",
       "      <td>0.691352</td>\n",
       "      <td>0.477968</td>\n",
       "      <td>4.488636</td>\n",
       "      <td>6.444131</td>\n",
       "      <td>6.884487</td>\n",
       "      <td>7.286192</td>\n",
       "      <td>10.105448</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>52</td>\n",
       "      <td>Chondromyces crocatus</td>\n",
       "      <td>UP000067626</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>8327</td>\n",
       "      <td>/ftp.uniprot.org/pub/databases/uniprot/current...</td>\n",
       "      <td>8327.0</td>\n",
       "      <td>403.466555</td>\n",
       "      <td>350.579560</td>\n",
       "      <td>122906.027917</td>\n",
       "      <td>29.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>486.0</td>\n",
       "      <td>8158.0</td>\n",
       "      <td>2.501095</td>\n",
       "      <td>0.300984</td>\n",
       "      <td>0.090591</td>\n",
       "      <td>1.462398</td>\n",
       "      <td>2.320146</td>\n",
       "      <td>2.513218</td>\n",
       "      <td>2.686636</td>\n",
       "      <td>3.911584</td>\n",
       "      <td>5.758984</td>\n",
       "      <td>0.693041</td>\n",
       "      <td>0.480306</td>\n",
       "      <td>3.367296</td>\n",
       "      <td>5.342334</td>\n",
       "      <td>5.786897</td>\n",
       "      <td>6.186209</td>\n",
       "      <td>9.006754</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>0.999080</td>\n",
       "      <td>-12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.998561</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         genes_species genes_assembly  \\\n",
       "0     archangium_gephyra_gca_001027285    ASM102728v1   \n",
       "1  chondromyces_crocatus_gca_001189295    ASM118929v1   \n",
       "\n",
       "                              genes_trunk_genes_path  \\\n",
       "0  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "1  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "\n",
       "                                    genes_genes_file  genes_count  \\\n",
       "0  protein_coding.genes.archangium_gephyra_gca_00...      10119.0   \n",
       "1  protein_coding.genes.chondromyces_crocatus_gca...       8339.0   \n",
       "\n",
       "    genes_mean    genes_std     genes_var  genes_min  genes_25perc  \\\n",
       "0  1103.871035  1127.066255  1.270278e+06      113.0         524.0   \n",
       "1  1211.514450  1051.408605  1.105460e+06       89.0         629.0   \n",
       "\n",
       "   genes_50perc  genes_75perc  genes_max  genes_log10_mean  genes_log10_std  \\\n",
       "0         887.0        1388.0    36653.0          2.919254         0.333869   \n",
       "1         977.0        1460.0    24476.0          2.978963         0.300250   \n",
       "\n",
       "   genes_log10_var  genes_log10_min  genes_log10_25perc  genes_log10_50perc  \\\n",
       "0         0.111468         2.053078            2.719331            2.947924   \n",
       "1         0.090150         1.949390            2.798651            2.989895   \n",
       "\n",
       "   genes_log10_75perc  genes_log10_max  genes_log_mean  genes_log_std  \\\n",
       "0            3.142389          4.56411        6.721831       0.768762   \n",
       "1            3.164353          4.38874        6.859315       0.691352   \n",
       "\n",
       "   genes_log_var  genes_log_min  genes_log_25perc  genes_log_50perc  \\\n",
       "0       0.590995       4.727388          6.261492          6.787845   \n",
       "1       0.477968       4.488636          6.444131          6.884487   \n",
       "\n",
       "   genes_log_75perc  genes_log_max genes_division  tax_id  \\\n",
       "0          7.235619      10.509251       bacteria      48   \n",
       "1          7.286192      10.105448       bacteria      52   \n",
       "\n",
       "           prots_species prots_proteome_id prots_superregnum  \\\n",
       "0     Archangium gephyra       UP000035579          bacteria   \n",
       "1  Chondromyces crocatus       UP000067626          bacteria   \n",
       "\n",
       "   prots_num_prot_cod_genes  \\\n",
       "0                     10110   \n",
       "1                      8327   \n",
       "\n",
       "                            prots_uniprot_fasta_file  prots_count  prots_mean  \\\n",
       "0  /ftp.uniprot.org/pub/databases/uniprot/current...      10110.0  367.499703   \n",
       "1  /ftp.uniprot.org/pub/databases/uniprot/current...       8327.0  403.466555   \n",
       "\n",
       "    prots_std      prots_var  prots_min  prots_25perc  prots_50perc  \\\n",
       "0  375.772244  141204.779058       37.0         174.0         295.0   \n",
       "1  350.579560  122906.027917       29.0         209.0         326.0   \n",
       "\n",
       "   prots_75perc  prots_max  prots_log10_mean  prots_log10_std  \\\n",
       "0         462.0    12217.0          2.441094         0.334836   \n",
       "1         486.0     8158.0          2.501095         0.300984   \n",
       "\n",
       "   prots_log10_var  prots_log10_min  prots_log10_25perc  prots_log10_50perc  \\\n",
       "0         0.112115         1.568202            2.240549            2.469822   \n",
       "1         0.090591         1.462398            2.320146            2.513218   \n",
       "\n",
       "   prots_log10_75perc  prots_log10_max  prots_log_mean  prots_log_std  \\\n",
       "0            2.664642         4.086965        5.620827       0.770987   \n",
       "1            2.686636         3.911584        5.758984       0.693041   \n",
       "\n",
       "   prots_log_var  prots_log_min  prots_log_25perc  prots_log_50perc  \\\n",
       "0       0.594422       3.610918          5.159055          5.686975   \n",
       "1       0.480306       3.367296          5.342334          5.786897   \n",
       "\n",
       "   prots_log_75perc  prots_log_max merged_division_superregnum  \\\n",
       "0          6.135565       9.410584                    bacteria   \n",
       "1          6.186209       9.006754                    bacteria   \n",
       "\n",
       "   ratio_mean_prots_genes  diff_prots_genes  abs_diff  ratio_prots_genes  \n",
       "0                0.998757              -9.0       9.0           0.999111  \n",
       "1                0.999080             -12.0      12.0           0.998561  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6519, 66)\n",
      "Index(['genes_species', 'genes_assembly', 'genes_trunk_genes_path',\n",
      "       'genes_genes_file', 'genes_count', 'genes_mean', 'genes_std',\n",
      "       'genes_var', 'genes_min', 'genes_25perc', 'genes_50perc',\n",
      "       'genes_75perc', 'genes_max', 'genes_log10_mean', 'genes_log10_std',\n",
      "       'genes_log10_var', 'genes_log10_min', 'genes_log10_25perc',\n",
      "       'genes_log10_50perc', 'genes_log10_75perc', 'genes_log10_max',\n",
      "       'genes_log_mean', 'genes_log_std', 'genes_log_var', 'genes_log_min',\n",
      "       'genes_log_25perc', 'genes_log_50perc', 'genes_log_75perc',\n",
      "       'genes_log_max', 'genes_division', 'tax_id', 'prots_species',\n",
      "       'prots_proteome_id', 'prots_superregnum', 'prots_num_prot_cod_genes',\n",
      "       'prots_uniprot_fasta_file', 'prots_count', 'prots_mean', 'prots_std',\n",
      "       'prots_var', 'prots_min', 'prots_25perc', 'prots_50perc',\n",
      "       'prots_75perc', 'prots_max', 'prots_log10_mean', 'prots_log10_std',\n",
      "       'prots_log10_var', 'prots_log10_min', 'prots_log10_25perc',\n",
      "       'prots_log10_50perc', 'prots_log10_75perc', 'prots_log10_max',\n",
      "       'prots_log_mean', 'prots_log_std', 'prots_log_var', 'prots_log_min',\n",
      "       'prots_log_25perc', 'prots_log_50perc', 'prots_log_75perc',\n",
      "       'prots_log_max', 'merged_division_superregnum',\n",
      "       'ratio_mean_prots_genes', 'diff_prots_genes', 'abs_diff',\n",
      "       'ratio_prots_genes'],\n",
      "      dtype='object')\n",
      "merged_division_superregnum\n",
      "bacteria         5468\n",
      "fungi             533\n",
      "archaea           227\n",
      "vertebrates        92\n",
      "protists           91\n",
      "plants             59\n",
      "invertebrates      49\n",
      "Name: count, dtype: int64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# statistics on length distribution for different species\n",
    "#########################################################\n",
    "#c.G_NCBI_GENOME_DATA_FILE #c.WRONG_ANNOTATIONS_MERGED_FILE #c.STAT_MERGED_FILE\n",
    "###############################################################################\n",
    "#\n",
    "stat_file = c.STAT_MERGED_FILE #c.G_NCBI_GENOME_DATA_FILE #c.WRONG_ANNOTATIONS_MERGED_FILE #c.STAT_MERGED_FILE\n",
    "out_file = stat_file\n",
    "print(\"The statistical descriptions of the protein distributions for the different species is in:\\n\", stat_file, \"\\n\")\n",
    "\n",
    "# retrieve data and diminish the number of columns\n",
    "stat_df = pd.read_csv(stat_file, low_memory=False, sep=\"\\t\")\n",
    "\n",
    "# visualize data\n",
    "if 1:\n",
    "    pd.set_option('display.max_columns', None)\n",
    "    display(stat_df.head(2))\n",
    "    print(stat_df.shape)\n",
    "    print(stat_df.columns)\n",
    "    if (stat_file == c.WRONG_ANNOTATIONS_MERGED_FILE) or (stat_file == c.STAT_MERGED_FILE):\n",
    "        print(stat_df[\"merged_division_superregnum\"].value_counts(), \"\\n\")\n",
    "    elif stat_file == c.G_NCBI_GENOME_DATA_FILE:  \n",
    "        print(stat_df[\"division_8\"].value_counts(), \"\\n\")\n",
    "    else:\n",
    "        sys.exit(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "759f7d0f-030f-467f-a000-2c8f597ba5d9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8e2ce8cc-0762-4e7f-bea5-a5cc57dd13f5",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Modify the data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a5f9daef-dd69-4971-b587-167e300dc7cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = stat_df.copy()\n",
    "if (stat_file == c.WRONG_ANNOTATIONS_MERGED_FILE) or (stat_file == c.STAT_MERGED_FILE):\n",
    "    df.loc[df.genes_division == \"metazoa\", 'genes_division'] = \"invertebrates\"\n",
    "    df.loc[df.merged_division_superregnum == \"metazoa\", 'merged_division_superregnum'] = \"invertebrates\"\n",
    "    df.loc[df.genes_division == \"protist\", 'genes_division'] = \"protists\"\n",
    "    df.loc[df.merged_division_superregnum == \"protist\", 'merged_division_superregnum'] = \"protists\"\n",
    "elif stat_file == c.G_NCBI_GENOME_DATA_FILE:    \n",
    "    df.loc[df.division_7 == \"metazoa\", 'division_7'] = \"invertebrates\"\n",
    "    df.loc[df.division_8 == \"metazoa\", 'division_8'] = \"invertebrates\"\n",
    "else:\n",
    "    sys.exit(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d6b86ab4-9b9d-473b-8b1a-cca8a384011b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes_species</th>\n",
       "      <th>genes_assembly</th>\n",
       "      <th>genes_trunk_genes_path</th>\n",
       "      <th>genes_genes_file</th>\n",
       "      <th>genes_count</th>\n",
       "      <th>genes_mean</th>\n",
       "      <th>genes_std</th>\n",
       "      <th>genes_var</th>\n",
       "      <th>genes_min</th>\n",
       "      <th>genes_25perc</th>\n",
       "      <th>genes_50perc</th>\n",
       "      <th>genes_75perc</th>\n",
       "      <th>genes_max</th>\n",
       "      <th>genes_log10_mean</th>\n",
       "      <th>genes_log10_std</th>\n",
       "      <th>genes_log10_var</th>\n",
       "      <th>genes_log10_min</th>\n",
       "      <th>genes_log10_25perc</th>\n",
       "      <th>genes_log10_50perc</th>\n",
       "      <th>genes_log10_75perc</th>\n",
       "      <th>genes_log10_max</th>\n",
       "      <th>genes_log_mean</th>\n",
       "      <th>genes_log_std</th>\n",
       "      <th>genes_log_var</th>\n",
       "      <th>genes_log_min</th>\n",
       "      <th>genes_log_25perc</th>\n",
       "      <th>genes_log_50perc</th>\n",
       "      <th>genes_log_75perc</th>\n",
       "      <th>genes_log_max</th>\n",
       "      <th>genes_division</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>prots_species</th>\n",
       "      <th>prots_proteome_id</th>\n",
       "      <th>prots_superregnum</th>\n",
       "      <th>prots_num_prot_cod_genes</th>\n",
       "      <th>prots_uniprot_fasta_file</th>\n",
       "      <th>prots_count</th>\n",
       "      <th>prots_mean</th>\n",
       "      <th>prots_std</th>\n",
       "      <th>prots_var</th>\n",
       "      <th>prots_min</th>\n",
       "      <th>prots_25perc</th>\n",
       "      <th>prots_50perc</th>\n",
       "      <th>prots_75perc</th>\n",
       "      <th>prots_max</th>\n",
       "      <th>prots_log10_mean</th>\n",
       "      <th>prots_log10_std</th>\n",
       "      <th>prots_log10_var</th>\n",
       "      <th>prots_log10_min</th>\n",
       "      <th>prots_log10_25perc</th>\n",
       "      <th>prots_log10_50perc</th>\n",
       "      <th>prots_log10_75perc</th>\n",
       "      <th>prots_log10_max</th>\n",
       "      <th>prots_log_mean</th>\n",
       "      <th>prots_log_std</th>\n",
       "      <th>prots_log_var</th>\n",
       "      <th>prots_log_min</th>\n",
       "      <th>prots_log_25perc</th>\n",
       "      <th>prots_log_50perc</th>\n",
       "      <th>prots_log_75perc</th>\n",
       "      <th>prots_log_max</th>\n",
       "      <th>merged_division_superregnum</th>\n",
       "      <th>ratio_mean_prots_genes</th>\n",
       "      <th>diff_prots_genes</th>\n",
       "      <th>abs_diff</th>\n",
       "      <th>ratio_prots_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>archangium_gephyra_gca_001027285</td>\n",
       "      <td>ASM102728v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.archangium_gephyra_gca_00...</td>\n",
       "      <td>10119.0</td>\n",
       "      <td>1103.871035</td>\n",
       "      <td>1127.066255</td>\n",
       "      <td>1.270278e+06</td>\n",
       "      <td>113.0</td>\n",
       "      <td>524.0</td>\n",
       "      <td>887.0</td>\n",
       "      <td>1388.0</td>\n",
       "      <td>36653.0</td>\n",
       "      <td>2.919254</td>\n",
       "      <td>0.333869</td>\n",
       "      <td>0.111468</td>\n",
       "      <td>2.053078</td>\n",
       "      <td>2.719331</td>\n",
       "      <td>2.947924</td>\n",
       "      <td>3.142389</td>\n",
       "      <td>4.56411</td>\n",
       "      <td>6.721831</td>\n",
       "      <td>0.768762</td>\n",
       "      <td>0.590995</td>\n",
       "      <td>4.727388</td>\n",
       "      <td>6.261492</td>\n",
       "      <td>6.787845</td>\n",
       "      <td>7.235619</td>\n",
       "      <td>10.509251</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>48</td>\n",
       "      <td>Archangium gephyra</td>\n",
       "      <td>UP000035579</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>10110</td>\n",
       "      <td>/ftp.uniprot.org/pub/databases/uniprot/current...</td>\n",
       "      <td>10110.0</td>\n",
       "      <td>367.499703</td>\n",
       "      <td>375.772244</td>\n",
       "      <td>141204.779058</td>\n",
       "      <td>37.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>295.0</td>\n",
       "      <td>462.0</td>\n",
       "      <td>12217.0</td>\n",
       "      <td>2.441094</td>\n",
       "      <td>0.334836</td>\n",
       "      <td>0.112115</td>\n",
       "      <td>1.568202</td>\n",
       "      <td>2.240549</td>\n",
       "      <td>2.469822</td>\n",
       "      <td>2.664642</td>\n",
       "      <td>4.086965</td>\n",
       "      <td>5.620827</td>\n",
       "      <td>0.770987</td>\n",
       "      <td>0.594422</td>\n",
       "      <td>3.610918</td>\n",
       "      <td>5.159055</td>\n",
       "      <td>5.686975</td>\n",
       "      <td>6.135565</td>\n",
       "      <td>9.410584</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>0.998757</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.999111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chondromyces_crocatus_gca_001189295</td>\n",
       "      <td>ASM118929v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.chondromyces_crocatus_gca...</td>\n",
       "      <td>8339.0</td>\n",
       "      <td>1211.514450</td>\n",
       "      <td>1051.408605</td>\n",
       "      <td>1.105460e+06</td>\n",
       "      <td>89.0</td>\n",
       "      <td>629.0</td>\n",
       "      <td>977.0</td>\n",
       "      <td>1460.0</td>\n",
       "      <td>24476.0</td>\n",
       "      <td>2.978963</td>\n",
       "      <td>0.300250</td>\n",
       "      <td>0.090150</td>\n",
       "      <td>1.949390</td>\n",
       "      <td>2.798651</td>\n",
       "      <td>2.989895</td>\n",
       "      <td>3.164353</td>\n",
       "      <td>4.38874</td>\n",
       "      <td>6.859315</td>\n",
       "      <td>0.691352</td>\n",
       "      <td>0.477968</td>\n",
       "      <td>4.488636</td>\n",
       "      <td>6.444131</td>\n",
       "      <td>6.884487</td>\n",
       "      <td>7.286192</td>\n",
       "      <td>10.105448</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>52</td>\n",
       "      <td>Chondromyces crocatus</td>\n",
       "      <td>UP000067626</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>8327</td>\n",
       "      <td>/ftp.uniprot.org/pub/databases/uniprot/current...</td>\n",
       "      <td>8327.0</td>\n",
       "      <td>403.466555</td>\n",
       "      <td>350.579560</td>\n",
       "      <td>122906.027917</td>\n",
       "      <td>29.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>486.0</td>\n",
       "      <td>8158.0</td>\n",
       "      <td>2.501095</td>\n",
       "      <td>0.300984</td>\n",
       "      <td>0.090591</td>\n",
       "      <td>1.462398</td>\n",
       "      <td>2.320146</td>\n",
       "      <td>2.513218</td>\n",
       "      <td>2.686636</td>\n",
       "      <td>3.911584</td>\n",
       "      <td>5.758984</td>\n",
       "      <td>0.693041</td>\n",
       "      <td>0.480306</td>\n",
       "      <td>3.367296</td>\n",
       "      <td>5.342334</td>\n",
       "      <td>5.786897</td>\n",
       "      <td>6.186209</td>\n",
       "      <td>9.006754</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>0.999080</td>\n",
       "      <td>-12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.998561</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         genes_species genes_assembly  \\\n",
       "0     archangium_gephyra_gca_001027285    ASM102728v1   \n",
       "1  chondromyces_crocatus_gca_001189295    ASM118929v1   \n",
       "\n",
       "                              genes_trunk_genes_path  \\\n",
       "0  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "1  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "\n",
       "                                    genes_genes_file  genes_count  \\\n",
       "0  protein_coding.genes.archangium_gephyra_gca_00...      10119.0   \n",
       "1  protein_coding.genes.chondromyces_crocatus_gca...       8339.0   \n",
       "\n",
       "    genes_mean    genes_std     genes_var  genes_min  genes_25perc  \\\n",
       "0  1103.871035  1127.066255  1.270278e+06      113.0         524.0   \n",
       "1  1211.514450  1051.408605  1.105460e+06       89.0         629.0   \n",
       "\n",
       "   genes_50perc  genes_75perc  genes_max  genes_log10_mean  genes_log10_std  \\\n",
       "0         887.0        1388.0    36653.0          2.919254         0.333869   \n",
       "1         977.0        1460.0    24476.0          2.978963         0.300250   \n",
       "\n",
       "   genes_log10_var  genes_log10_min  genes_log10_25perc  genes_log10_50perc  \\\n",
       "0         0.111468         2.053078            2.719331            2.947924   \n",
       "1         0.090150         1.949390            2.798651            2.989895   \n",
       "\n",
       "   genes_log10_75perc  genes_log10_max  genes_log_mean  genes_log_std  \\\n",
       "0            3.142389          4.56411        6.721831       0.768762   \n",
       "1            3.164353          4.38874        6.859315       0.691352   \n",
       "\n",
       "   genes_log_var  genes_log_min  genes_log_25perc  genes_log_50perc  \\\n",
       "0       0.590995       4.727388          6.261492          6.787845   \n",
       "1       0.477968       4.488636          6.444131          6.884487   \n",
       "\n",
       "   genes_log_75perc  genes_log_max genes_division  tax_id  \\\n",
       "0          7.235619      10.509251       bacteria      48   \n",
       "1          7.286192      10.105448       bacteria      52   \n",
       "\n",
       "           prots_species prots_proteome_id prots_superregnum  \\\n",
       "0     Archangium gephyra       UP000035579          bacteria   \n",
       "1  Chondromyces crocatus       UP000067626          bacteria   \n",
       "\n",
       "   prots_num_prot_cod_genes  \\\n",
       "0                     10110   \n",
       "1                      8327   \n",
       "\n",
       "                            prots_uniprot_fasta_file  prots_count  prots_mean  \\\n",
       "0  /ftp.uniprot.org/pub/databases/uniprot/current...      10110.0  367.499703   \n",
       "1  /ftp.uniprot.org/pub/databases/uniprot/current...       8327.0  403.466555   \n",
       "\n",
       "    prots_std      prots_var  prots_min  prots_25perc  prots_50perc  \\\n",
       "0  375.772244  141204.779058       37.0         174.0         295.0   \n",
       "1  350.579560  122906.027917       29.0         209.0         326.0   \n",
       "\n",
       "   prots_75perc  prots_max  prots_log10_mean  prots_log10_std  \\\n",
       "0         462.0    12217.0          2.441094         0.334836   \n",
       "1         486.0     8158.0          2.501095         0.300984   \n",
       "\n",
       "   prots_log10_var  prots_log10_min  prots_log10_25perc  prots_log10_50perc  \\\n",
       "0         0.112115         1.568202            2.240549            2.469822   \n",
       "1         0.090591         1.462398            2.320146            2.513218   \n",
       "\n",
       "   prots_log10_75perc  prots_log10_max  prots_log_mean  prots_log_std  \\\n",
       "0            2.664642         4.086965        5.620827       0.770987   \n",
       "1            2.686636         3.911584        5.758984       0.693041   \n",
       "\n",
       "   prots_log_var  prots_log_min  prots_log_25perc  prots_log_50perc  \\\n",
       "0       0.594422       3.610918          5.159055          5.686975   \n",
       "1       0.480306       3.367296          5.342334          5.786897   \n",
       "\n",
       "   prots_log_75perc  prots_log_max merged_division_superregnum  \\\n",
       "0          6.135565       9.410584                    bacteria   \n",
       "1          6.186209       9.006754                    bacteria   \n",
       "\n",
       "   ratio_mean_prots_genes  diff_prots_genes  abs_diff  ratio_prots_genes  \n",
       "0                0.998757              -9.0       9.0           0.999111  \n",
       "1                0.999080             -12.0      12.0           0.998561  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6519, 66)\n",
      "Index(['genes_species', 'genes_assembly', 'genes_trunk_genes_path',\n",
      "       'genes_genes_file', 'genes_count', 'genes_mean', 'genes_std',\n",
      "       'genes_var', 'genes_min', 'genes_25perc', 'genes_50perc',\n",
      "       'genes_75perc', 'genes_max', 'genes_log10_mean', 'genes_log10_std',\n",
      "       'genes_log10_var', 'genes_log10_min', 'genes_log10_25perc',\n",
      "       'genes_log10_50perc', 'genes_log10_75perc', 'genes_log10_max',\n",
      "       'genes_log_mean', 'genes_log_std', 'genes_log_var', 'genes_log_min',\n",
      "       'genes_log_25perc', 'genes_log_50perc', 'genes_log_75perc',\n",
      "       'genes_log_max', 'genes_division', 'tax_id', 'prots_species',\n",
      "       'prots_proteome_id', 'prots_superregnum', 'prots_num_prot_cod_genes',\n",
      "       'prots_uniprot_fasta_file', 'prots_count', 'prots_mean', 'prots_std',\n",
      "       'prots_var', 'prots_min', 'prots_25perc', 'prots_50perc',\n",
      "       'prots_75perc', 'prots_max', 'prots_log10_mean', 'prots_log10_std',\n",
      "       'prots_log10_var', 'prots_log10_min', 'prots_log10_25perc',\n",
      "       'prots_log10_50perc', 'prots_log10_75perc', 'prots_log10_max',\n",
      "       'prots_log_mean', 'prots_log_std', 'prots_log_var', 'prots_log_min',\n",
      "       'prots_log_25perc', 'prots_log_50perc', 'prots_log_75perc',\n",
      "       'prots_log_max', 'merged_division_superregnum',\n",
      "       'ratio_mean_prots_genes', 'diff_prots_genes', 'abs_diff',\n",
      "       'ratio_prots_genes'],\n",
      "      dtype='object')\n",
      "merged_division_superregnum\n",
      "bacteria         5468\n",
      "fungi             533\n",
      "archaea           227\n",
      "vertebrates        92\n",
      "protists           91\n",
      "plants             59\n",
      "invertebrates      49\n",
      "Name: count, dtype: int64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "if 1:\n",
    "    pd.set_option('display.max_columns', None)\n",
    "    display(stat_df.head(2))\n",
    "    print(df.shape)\n",
    "    print(df.columns)\n",
    "    if (stat_file == c.WRONG_ANNOTATIONS_MERGED_FILE) or (stat_file == c.STAT_MERGED_FILE):\n",
    "        print(df[\"merged_division_superregnum\"].value_counts(), \"\\n\")\n",
    "    elif stat_file == c.G_NCBI_GENOME_DATA_FILE:    \n",
    "        print(df[\"division_8\"].value_counts(), \"\\n\")\n",
    "    else:\n",
    "        sys.exit(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d177f840-5a83-41be-9ec4-21f143b8dc48",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f3bf5d5a-a73c-4c74-ac86-a059f93775d6",
   "metadata": {},
   "source": [
    "## Save file if desired\n",
    "be careful"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4ab100fc-65ed-41cd-9746-e6d193b2b7a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/emuro/git/github/EM_geneLength_nature/working_on/preprocessing_data/../../lib/../main_tables/stat_merged.tsv True\n"
     ]
    }
   ],
   "source": [
    "print(out_file, os.path.isfile(out_file))\n",
    "if BOOL_SAVE_FILE: # Save the file if desired\n",
    "    print(\"saving in: \", out_file)\n",
    "    df.to_csv(out_file, sep=\"\\t\", index=False, header=True)\n",
    "    print(\"...file saved. Done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc4f7d83-1344-43ab-bba0-61c44ffee1bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
