{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "819bd65b-f5b5-4644-a058-0a8c843d4fe8",
   "metadata": {},
   "source": [
    "# Distribution of the lengths of the protein coding genes \n",
    "**for the different species**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8046a185-31f3-48b1-83f7-947911a96d15",
   "metadata": {},
   "source": [
    "## Import python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "51b0a90c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "from scipy import stats \n",
    "import os\n",
    "import sys\n",
    "sys.path.append('../../lib/')\n",
    "import constants as c\n",
    "\n",
    "\n",
    "BOOL_SAVE_FILE = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7b6d3b5",
   "metadata": {},
   "source": [
    "## The protein coding gene length distributions for the different species"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d509543a-9bbf-4428-bb98-a287201ef79d",
   "metadata": {},
   "source": [
    "### Retrieving the statistical descriptions of the gene length's distributions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a99f2a8-5073-440f-88be-bf0def58ea9d",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Statistical description "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e96ab72e-e702-4020-814a-625c86dd44ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The statistical descriptions of the protein coding gene distributions for the different species is in:\n",
      " /home/emuro/git/github/EM_geneLength_nature/working_on/preprocessing_data/../../lib/../main_tables/stat_protCodGenes.tsv \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_original</th>\n",
       "      <th>division_7</th>\n",
       "      <th>species</th>\n",
       "      <th>assembly</th>\n",
       "      <th>trunk_genes_path</th>\n",
       "      <th>genes_file</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>var</th>\n",
       "      <th>min</th>\n",
       "      <th>25perc</th>\n",
       "      <th>50perc</th>\n",
       "      <th>75perc</th>\n",
       "      <th>max</th>\n",
       "      <th>log10_mean</th>\n",
       "      <th>log10_std</th>\n",
       "      <th>log10_var</th>\n",
       "      <th>log10_min</th>\n",
       "      <th>log10_25perc</th>\n",
       "      <th>log10_50perc</th>\n",
       "      <th>log10_75perc</th>\n",
       "      <th>log10_max</th>\n",
       "      <th>log_mean</th>\n",
       "      <th>log_std</th>\n",
       "      <th>log_var</th>\n",
       "      <th>log_min</th>\n",
       "      <th>log_25perc</th>\n",
       "      <th>log_50perc</th>\n",
       "      <th>log_75perc</th>\n",
       "      <th>log_max</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>genebuild</th>\n",
       "      <th>Scientific_name</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>division_4_byLineage</th>\n",
       "      <th>Virus_host</th>\n",
       "      <th>division_8</th>\n",
       "      <th>bool_Rank</th>\n",
       "      <th>sameTax_order</th>\n",
       "      <th>division_both_dbs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7068</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>methanobacterium_bryantii_gca_002287175</td>\n",
       "      <td>ASM228717v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.methanobacterium_bryantii...</td>\n",
       "      <td>3168</td>\n",
       "      <td>840.404040</td>\n",
       "      <td>649.879873</td>\n",
       "      <td>422343.848699</td>\n",
       "      <td>131</td>\n",
       "      <td>416.0</td>\n",
       "      <td>686.0</td>\n",
       "      <td>1079.0</td>\n",
       "      <td>7910</td>\n",
       "      <td>2.827684</td>\n",
       "      <td>0.287565</td>\n",
       "      <td>0.082694</td>\n",
       "      <td>2.117271</td>\n",
       "      <td>2.619093</td>\n",
       "      <td>2.836324</td>\n",
       "      <td>3.033021</td>\n",
       "      <td>3.898176</td>\n",
       "      <td>6.510982</td>\n",
       "      <td>0.662143</td>\n",
       "      <td>0.438433</td>\n",
       "      <td>4.875197</td>\n",
       "      <td>6.030685</td>\n",
       "      <td>6.530878</td>\n",
       "      <td>6.98379</td>\n",
       "      <td>8.975883</td>\n",
       "      <td>2161</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>Methanobacterium bryantii</td>\n",
       "      <td>Species</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>archaea</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>archaea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26156</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>methanobacterium_formicicum_gca_000762265</td>\n",
       "      <td>ASM76226v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.methanobacterium_formicic...</td>\n",
       "      <td>2352</td>\n",
       "      <td>862.427296</td>\n",
       "      <td>589.270407</td>\n",
       "      <td>347239.612747</td>\n",
       "      <td>107</td>\n",
       "      <td>458.0</td>\n",
       "      <td>728.0</td>\n",
       "      <td>1115.0</td>\n",
       "      <td>6779</td>\n",
       "      <td>2.849723</td>\n",
       "      <td>0.276255</td>\n",
       "      <td>0.076317</td>\n",
       "      <td>2.029384</td>\n",
       "      <td>2.660865</td>\n",
       "      <td>2.862131</td>\n",
       "      <td>3.047275</td>\n",
       "      <td>3.831166</td>\n",
       "      <td>6.561730</td>\n",
       "      <td>0.636100</td>\n",
       "      <td>0.404623</td>\n",
       "      <td>4.672829</td>\n",
       "      <td>6.126869</td>\n",
       "      <td>6.590301</td>\n",
       "      <td>7.01661</td>\n",
       "      <td>8.821585</td>\n",
       "      <td>2162</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>Methanobacterium formicicum</td>\n",
       "      <td>Species</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>archaea</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>archaea</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index_original division_7                                    species  \\\n",
       "0            7068   bacteria    methanobacterium_bryantii_gca_002287175   \n",
       "1           26156   bacteria  methanobacterium_formicicum_gca_000762265   \n",
       "\n",
       "      assembly                                   trunk_genes_path  \\\n",
       "0  ASM228717v1  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "1   ASM76226v1  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "\n",
       "                                          genes_file  count        mean  \\\n",
       "0  protein_coding.genes.methanobacterium_bryantii...   3168  840.404040   \n",
       "1  protein_coding.genes.methanobacterium_formicic...   2352  862.427296   \n",
       "\n",
       "          std            var  min  25perc  50perc  75perc   max  log10_mean  \\\n",
       "0  649.879873  422343.848699  131   416.0   686.0  1079.0  7910    2.827684   \n",
       "1  589.270407  347239.612747  107   458.0   728.0  1115.0  6779    2.849723   \n",
       "\n",
       "   log10_std  log10_var  log10_min  log10_25perc  log10_50perc  log10_75perc  \\\n",
       "0   0.287565   0.082694   2.117271      2.619093      2.836324      3.033021   \n",
       "1   0.276255   0.076317   2.029384      2.660865      2.862131      3.047275   \n",
       "\n",
       "   log10_max  log_mean   log_std   log_var   log_min  log_25perc  log_50perc  \\\n",
       "0   3.898176  6.510982  0.662143  0.438433  4.875197    6.030685    6.530878   \n",
       "1   3.831166  6.561730  0.636100  0.404623  4.672829    6.126869    6.590301   \n",
       "\n",
       "   log_75perc   log_max  taxonomy_id  genebuild              Scientific_name  \\\n",
       "0     6.98379  8.975883         2161     2017.0    Methanobacterium bryantii   \n",
       "1     7.01661  8.821585         2162     2014.0  Methanobacterium formicicum   \n",
       "\n",
       "      Rank                                            Lineage  \\\n",
       "0  Species  Archaea; Euryarchaeota; Methanomada group; Met...   \n",
       "1  Species  Archaea; Euryarchaeota; Methanomada group; Met...   \n",
       "\n",
       "  division_4_byLineage  Virus_host division_8  bool_Rank  sameTax_order  \\\n",
       "0              Archaea         NaN    archaea          1              1   \n",
       "1              Archaea         NaN    archaea          1              1   \n",
       "\n",
       "  division_both_dbs  \n",
       "0           archaea  \n",
       "1           archaea  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(33627, 42)\n",
      "Index(['index_original', 'division_7', 'species', 'assembly',\n",
      "       'trunk_genes_path', 'genes_file', 'count', 'mean', 'std', 'var', 'min',\n",
      "       '25perc', '50perc', '75perc', 'max', 'log10_mean', 'log10_std',\n",
      "       'log10_var', 'log10_min', 'log10_25perc', 'log10_50perc',\n",
      "       'log10_75perc', 'log10_max', 'log_mean', 'log_std', 'log_var',\n",
      "       'log_min', 'log_25perc', 'log_50perc', 'log_75perc', 'log_max',\n",
      "       'taxonomy_id', 'genebuild', 'Scientific_name', 'Rank', 'Lineage',\n",
      "       'division_4_byLineage', 'Virus_host', 'division_8', 'bool_Rank',\n",
      "       'sameTax_order', 'division_both_dbs'],\n",
      "      dtype='object')\n",
      "division_7\n",
      "bacteria         31943\n",
      "fungi             1014\n",
      "protists           237\n",
      "vertebrates        222\n",
      "invertebrates      115\n",
      "plants              96\n",
      "Name: count, dtype: int64 \n",
      "\n",
      "division_8\n",
      "bacteria         30555\n",
      "archaea           1228\n",
      "fungi             1007\n",
      "protists           237\n",
      "vertebrates        221\n",
      "invertebrates      115\n",
      "plants              96\n",
      "Name: count, dtype: int64 \n",
      "\n",
      "division_both_dbs\n",
      "bacteria         30714\n",
      "archaea           1229\n",
      "fungi             1014\n",
      "protists           237\n",
      "vertebrates        222\n",
      "invertebrates      115\n",
      "plants              96\n",
      "Name: count, dtype: int64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# statistics on length distributions \n",
    "stat_file = c.STAT_G_FILE\n",
    "print(\"The statistical descriptions of the protein coding gene distributions for the different species is in:\\n\", stat_file, \"\\n\")\n",
    "\n",
    "# retrieve data and diminish the number of columns\n",
    "stat_df = pd.read_csv(stat_file, low_memory=False, sep=\"\\t\")\n",
    "\n",
    "# visualize data\n",
    "if 1:\n",
    "    pd.set_option('display.max_columns', None)\n",
    "    display(stat_df.head(2))\n",
    "    print(stat_df.shape)\n",
    "    print(stat_df.columns)\n",
    "    print(stat_df[\"division_7\"].value_counts(), \"\\n\")\n",
    "    print(stat_df[\"division_8\"].value_counts(), \"\\n\")\n",
    "    print(stat_df[\"division_both_dbs\"].value_counts(), \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7864e7f-66e1-4115-9a45-c2d020171688",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Modify the data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a5f9daef-dd69-4971-b587-167e300dc7cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = stat_df.copy()\n",
    "df.loc[df.division_7 == \"metazoa\",        'division_7']        = \"invertebrates\"\n",
    "df.loc[df.division_8 == \"metazoa\",        'division_8']        = \"invertebrates\"\n",
    "df.loc[df.division_both_dbs == \"metazoa\", 'division_both_dbs'] = \"invertebrates\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d6b86ab4-9b9d-473b-8b1a-cca8a384011b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_original</th>\n",
       "      <th>division_7</th>\n",
       "      <th>species</th>\n",
       "      <th>assembly</th>\n",
       "      <th>trunk_genes_path</th>\n",
       "      <th>genes_file</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>var</th>\n",
       "      <th>min</th>\n",
       "      <th>25perc</th>\n",
       "      <th>50perc</th>\n",
       "      <th>75perc</th>\n",
       "      <th>max</th>\n",
       "      <th>log10_mean</th>\n",
       "      <th>log10_std</th>\n",
       "      <th>log10_var</th>\n",
       "      <th>log10_min</th>\n",
       "      <th>log10_25perc</th>\n",
       "      <th>log10_50perc</th>\n",
       "      <th>log10_75perc</th>\n",
       "      <th>log10_max</th>\n",
       "      <th>log_mean</th>\n",
       "      <th>log_std</th>\n",
       "      <th>log_var</th>\n",
       "      <th>log_min</th>\n",
       "      <th>log_25perc</th>\n",
       "      <th>log_50perc</th>\n",
       "      <th>log_75perc</th>\n",
       "      <th>log_max</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>genebuild</th>\n",
       "      <th>Scientific_name</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>division_4_byLineage</th>\n",
       "      <th>Virus_host</th>\n",
       "      <th>division_8</th>\n",
       "      <th>bool_Rank</th>\n",
       "      <th>sameTax_order</th>\n",
       "      <th>division_both_dbs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7068</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>methanobacterium_bryantii_gca_002287175</td>\n",
       "      <td>ASM228717v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.methanobacterium_bryantii...</td>\n",
       "      <td>3168</td>\n",
       "      <td>840.404040</td>\n",
       "      <td>649.879873</td>\n",
       "      <td>422343.848699</td>\n",
       "      <td>131</td>\n",
       "      <td>416.0</td>\n",
       "      <td>686.0</td>\n",
       "      <td>1079.0</td>\n",
       "      <td>7910</td>\n",
       "      <td>2.827684</td>\n",
       "      <td>0.287565</td>\n",
       "      <td>0.082694</td>\n",
       "      <td>2.117271</td>\n",
       "      <td>2.619093</td>\n",
       "      <td>2.836324</td>\n",
       "      <td>3.033021</td>\n",
       "      <td>3.898176</td>\n",
       "      <td>6.510982</td>\n",
       "      <td>0.662143</td>\n",
       "      <td>0.438433</td>\n",
       "      <td>4.875197</td>\n",
       "      <td>6.030685</td>\n",
       "      <td>6.530878</td>\n",
       "      <td>6.98379</td>\n",
       "      <td>8.975883</td>\n",
       "      <td>2161</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>Methanobacterium bryantii</td>\n",
       "      <td>Species</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>archaea</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>archaea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26156</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>methanobacterium_formicicum_gca_000762265</td>\n",
       "      <td>ASM76226v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.methanobacterium_formicic...</td>\n",
       "      <td>2352</td>\n",
       "      <td>862.427296</td>\n",
       "      <td>589.270407</td>\n",
       "      <td>347239.612747</td>\n",
       "      <td>107</td>\n",
       "      <td>458.0</td>\n",
       "      <td>728.0</td>\n",
       "      <td>1115.0</td>\n",
       "      <td>6779</td>\n",
       "      <td>2.849723</td>\n",
       "      <td>0.276255</td>\n",
       "      <td>0.076317</td>\n",
       "      <td>2.029384</td>\n",
       "      <td>2.660865</td>\n",
       "      <td>2.862131</td>\n",
       "      <td>3.047275</td>\n",
       "      <td>3.831166</td>\n",
       "      <td>6.561730</td>\n",
       "      <td>0.636100</td>\n",
       "      <td>0.404623</td>\n",
       "      <td>4.672829</td>\n",
       "      <td>6.126869</td>\n",
       "      <td>6.590301</td>\n",
       "      <td>7.01661</td>\n",
       "      <td>8.821585</td>\n",
       "      <td>2162</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>Methanobacterium formicicum</td>\n",
       "      <td>Species</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>archaea</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>archaea</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index_original division_7                                    species  \\\n",
       "0            7068   bacteria    methanobacterium_bryantii_gca_002287175   \n",
       "1           26156   bacteria  methanobacterium_formicicum_gca_000762265   \n",
       "\n",
       "      assembly                                   trunk_genes_path  \\\n",
       "0  ASM228717v1  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "1   ASM76226v1  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "\n",
       "                                          genes_file  count        mean  \\\n",
       "0  protein_coding.genes.methanobacterium_bryantii...   3168  840.404040   \n",
       "1  protein_coding.genes.methanobacterium_formicic...   2352  862.427296   \n",
       "\n",
       "          std            var  min  25perc  50perc  75perc   max  log10_mean  \\\n",
       "0  649.879873  422343.848699  131   416.0   686.0  1079.0  7910    2.827684   \n",
       "1  589.270407  347239.612747  107   458.0   728.0  1115.0  6779    2.849723   \n",
       "\n",
       "   log10_std  log10_var  log10_min  log10_25perc  log10_50perc  log10_75perc  \\\n",
       "0   0.287565   0.082694   2.117271      2.619093      2.836324      3.033021   \n",
       "1   0.276255   0.076317   2.029384      2.660865      2.862131      3.047275   \n",
       "\n",
       "   log10_max  log_mean   log_std   log_var   log_min  log_25perc  log_50perc  \\\n",
       "0   3.898176  6.510982  0.662143  0.438433  4.875197    6.030685    6.530878   \n",
       "1   3.831166  6.561730  0.636100  0.404623  4.672829    6.126869    6.590301   \n",
       "\n",
       "   log_75perc   log_max  taxonomy_id  genebuild              Scientific_name  \\\n",
       "0     6.98379  8.975883         2161     2017.0    Methanobacterium bryantii   \n",
       "1     7.01661  8.821585         2162     2014.0  Methanobacterium formicicum   \n",
       "\n",
       "      Rank                                            Lineage  \\\n",
       "0  Species  Archaea; Euryarchaeota; Methanomada group; Met...   \n",
       "1  Species  Archaea; Euryarchaeota; Methanomada group; Met...   \n",
       "\n",
       "  division_4_byLineage  Virus_host division_8  bool_Rank  sameTax_order  \\\n",
       "0              Archaea         NaN    archaea          1              1   \n",
       "1              Archaea         NaN    archaea          1              1   \n",
       "\n",
       "  division_both_dbs  \n",
       "0           archaea  \n",
       "1           archaea  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(33627, 42)\n",
      "Index(['index_original', 'division_7', 'species', 'assembly',\n",
      "       'trunk_genes_path', 'genes_file', 'count', 'mean', 'std', 'var', 'min',\n",
      "       '25perc', '50perc', '75perc', 'max', 'log10_mean', 'log10_std',\n",
      "       'log10_var', 'log10_min', 'log10_25perc', 'log10_50perc',\n",
      "       'log10_75perc', 'log10_max', 'log_mean', 'log_std', 'log_var',\n",
      "       'log_min', 'log_25perc', 'log_50perc', 'log_75perc', 'log_max',\n",
      "       'taxonomy_id', 'genebuild', 'Scientific_name', 'Rank', 'Lineage',\n",
      "       'division_4_byLineage', 'Virus_host', 'division_8', 'bool_Rank',\n",
      "       'sameTax_order', 'division_both_dbs'],\n",
      "      dtype='object')\n",
      "division_7\n",
      "bacteria         31943\n",
      "fungi             1014\n",
      "protists           237\n",
      "vertebrates        222\n",
      "invertebrates      115\n",
      "plants              96\n",
      "Name: count, dtype: int64 \n",
      "\n",
      "division_8\n",
      "bacteria         30555\n",
      "archaea           1228\n",
      "fungi             1007\n",
      "protists           237\n",
      "vertebrates        221\n",
      "invertebrates      115\n",
      "plants              96\n",
      "Name: count, dtype: int64 \n",
      "\n",
      "division_both_dbs\n",
      "bacteria         30714\n",
      "archaea           1229\n",
      "fungi             1014\n",
      "protists           237\n",
      "vertebrates        222\n",
      "invertebrates      115\n",
      "plants              96\n",
      "Name: count, dtype: int64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "if 1:\n",
    "    pd.set_option('display.max_columns', None)\n",
    "    display(stat_df.head(2))\n",
    "    print(df.shape)\n",
    "    print(df.columns)\n",
    "    print(df[\"division_7\"].value_counts(), \"\\n\")\n",
    "    print(df[\"division_8\"].value_counts(), \"\\n\")\n",
    "    print(df[\"division_both_dbs\"].value_counts(), \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a1d6df9-b7fd-49a1-85f8-6006e9d66f60",
   "metadata": {},
   "source": [
    "## Save file if desired\n",
    "be careful"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4ab100fc-65ed-41cd-9746-e6d193b2b7a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/emuro/git/github/EM_geneLength_nature/working_on/preprocessing_data/../../lib/../main_tables/stat_protCodGenes.tsv True\n"
     ]
    }
   ],
   "source": [
    "out_file = c.STAT_G_FILE\n",
    "print(out_file, os.path.isfile(out_file))\n",
    "if BOOL_SAVE_FILE: # Save the file if desired\n",
    "    print(\"saving in: \", out_file)\n",
    "    df.to_csv(out_file, sep=\"\\t\", index=False, header=True)\n",
    "    print(\"...file saved. Done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc4f7d83-1344-43ab-bba0-61c44ffee1bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
