{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "819bd65b-f5b5-4644-a058-0a8c843d4fe8",
   "metadata": {},
   "source": [
    "# Observed Taylor law: variance vs mean \n",
    "**in protein coding gene length distributions for the different species**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8046a185-31f3-48b1-83f7-947911a96d15",
   "metadata": {},
   "source": [
    "## Import python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "51b0a90c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "#import matplotlib.pyplot as plt\n",
    "from scipy import stats\n",
    "#import os \n",
    "import pandas as pd\n",
    "#import math\n",
    "#\n",
    "from plotnine import *\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d509543a-9bbf-4428-bb98-a287201ef79d",
   "metadata": {},
   "source": [
    "## Retrieving the statistical descriptions of the gene length's distributions for the different genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e96ab72e-e702-4020-814a-625c86dd44ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The statistical descriptions of the protein coding gene distributions for the different species is in:\n",
      " ../../../main_tables/stat_protCodGenes.tsv \n",
      "\n",
      "(33627, 6)\n",
      "division_7\n",
      "bacteria       31943\n",
      "fungi           1014\n",
      "protists         237\n",
      "vertebrates      222\n",
      "metazoa          115\n",
      "plants            96\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>division_7</th>\n",
       "      <th>division_8</th>\n",
       "      <th>species</th>\n",
       "      <th>assembly</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>Lineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>archaea</td>\n",
       "      <td>methanobacterium_bryantii_gca_002287175</td>\n",
       "      <td>ASM228717v1</td>\n",
       "      <td>2161</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>archaea</td>\n",
       "      <td>methanobacterium_formicicum_gca_000762265</td>\n",
       "      <td>ASM76226v1</td>\n",
       "      <td>2162</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  division_7 division_8                                    species  \\\n",
       "0   bacteria    archaea    methanobacterium_bryantii_gca_002287175   \n",
       "1   bacteria    archaea  methanobacterium_formicicum_gca_000762265   \n",
       "\n",
       "      assembly  taxonomy_id                                            Lineage  \n",
       "0  ASM228717v1         2161  Archaea; Euryarchaeota; Methanomada group; Met...  \n",
       "1   ASM76226v1         2162  Archaea; Euryarchaeota; Methanomada group; Met...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# statistics on length distribution for different species\n",
    "working_on_extra_path = \"../../../\"\n",
    "stat_file = working_on_extra_path + \"main_tables/stat_protCodGenes.tsv\"\n",
    "print(\"The statistical descriptions of the protein coding gene distributions for the different species is in:\\n\", stat_file, \"\\n\")\n",
    "\n",
    "# retrieve data and diminish the number of columns\n",
    "stat_df = pd.read_csv(stat_file, low_memory=False, sep=\"\\t\")\n",
    "stat_df = stat_df[[\"division_7\", \"division_8\", \"species\", \"assembly\", \"taxonomy_id\", \"Lineage\"]] \n",
    "\n",
    "# visualize data\n",
    "pd.set_option('display.max_columns', None)\n",
    "if 1:\n",
    "    print(stat_df.shape)\n",
    "    print(stat_df.value_counts('division_7', dropna=False))\n",
    "    display(stat_df.head(2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7864e7f-66e1-4115-9a45-c2d020171688",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Python functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3d2a2be-dadb-4ba4-8aee-3b3f457a2e0d",
   "metadata": {},
   "source": [
    "### All protein coding genes from the paper (33,627)\n",
    "incluyendo 168 entradas de division_7 que se pierden al hacer el label division_8."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "773e6141-e5fe-4c4e-85d7-47c558251acc",
   "metadata": {},
   "source": [
    "## Filtro por calidad de Genomas: usando NCBI genome annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2bb61a2-bf47-4fe1-bace-820ff542833c",
   "metadata": {},
   "source": [
    "#### Get good genomes\n",
    "Genome status: Complete Genome or Chromosomes (from NCBI genome annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f432606b-2d34-411a-bd85-1d70d5f68d50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/emuro\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "home = str(Path.home())\n",
    "print(home)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "10dfb213-afc8-4118-8f61-81f729c332f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "prokaryotes (405092, 23)\n",
      "eukaryotes (22897, 19)\n",
      "prokaryotes (405092, 23)\n",
      "eukaryotes (22897, 19)\n"
     ]
    }
   ],
   "source": [
    "# Prokarya genomes: get those with a good genome annotation (\"Status\")\n",
    "#\n",
    "ncbi_genome_home = home + \"/Desktop/goingOn/geneLength/NCBI_genomeReports/\"\n",
    "col1 = \"TaxID\"\n",
    "col2 = \"Assembly Accession\"\n",
    "\n",
    "good_status = [\"Complete Genome\", \"Chromosome\"] #[\"Complete Genome\", \"Chromosome\", \"Scaffold\", \"Contig\"]\n",
    "\n",
    "# Get well annotated eukarya genomes\n",
    "g_prok_df = pd.read_csv(ncbi_genome_home + \"prokaryotes.txt\", low_memory=False, sep=\"\\t\") # tax_id, status, accession \n",
    "g_euk_df  = pd.read_csv(ncbi_genome_home + \"eukaryotes.txt\", sep=\"\\t\")                    \n",
    "print(\"prokaryotes\", g_prok_df.shape) \n",
    "print(\"eukaryotes\",  g_euk_df.shape)\n",
    "\n",
    "# Ya filtra la calidad!\n",
    "if 0: # 0: no filtra por calidad\n",
    "    g_prok_df = g_prok_df[ g_prok_df[\"Status\"].isin(good_status)]\n",
    "    g_euk_df  = g_euk_df[  g_euk_df[\"Status\"].isin(good_status)]\n",
    "print(\"prokaryotes\", g_prok_df.shape)\n",
    "print(\"eukaryotes\", g_euk_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ca3e7ea-ce02-4ffc-bc3c-829e64305e4f",
   "metadata": {},
   "source": [
    "### From NCBI Assemblies files \n",
    "- dict: GC\\% of Assembly_accession of species\n",
    "- dict: size (Mbp) of Assembly_accession\n",
    "- dict: replicons of Assembly_accession\n",
    "- dict: status of Assembly_accession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bf113e0c-43cd-4529-a0fc-ab4936f562b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Only high quality genomes!..If previously filtered\n",
      "22897 Eukaryotes\n",
      "405092 Prokaryotes\n",
      "427989 Eukaryotes & Prokaryotes\n"
     ]
    }
   ],
   "source": [
    "print(\"Only high quality genomes!..If previously filtered\")\n",
    "\n",
    "#\n",
    "# Eukaryotes\n",
    "if 0:\n",
    "    print(g_euk_df.columns)\n",
    "dict_euk_groupOfAssembly     = dict(zip(g_euk_df[\"Assembly Accession\"], g_euk_df[\"Group\"]))         \n",
    "dict_euk_subgroupOfAssembly  = dict(zip(g_euk_df[\"Assembly Accession\"], g_euk_df[\"SubGroup\"]))    \n",
    "dict_euk_taxidOfAssembly     = dict(zip(g_euk_df[\"Assembly Accession\"], g_euk_df[\"TaxID\"]))          # print(dict_euk_taxidOfAssembly)\n",
    "dict_euk_gcOfAssembly        = dict(zip(g_euk_df[\"Assembly Accession\"], g_euk_df[\"GC%\"]))            \n",
    "dict_euk_sizeOfAssembly      = dict(zip(g_euk_df[\"Assembly Accession\"], g_euk_df[\"Size (Mb)\"]))      \n",
    "dict_euk_repliconsOfAssembly = dict(zip(g_euk_df[\"Assembly Accession\"], g_euk_df[\"Replicons\"]))     \n",
    "dict_euk_statusOfAssembly    = dict(zip(g_euk_df[\"Assembly Accession\"], g_euk_df[\"Status\"]))         \n",
    "if 1:\n",
    "  print(len(dict_euk_statusOfAssembly), \"Eukaryotes\")\n",
    "\n",
    "# Prokaryotes\n",
    "if 0:\n",
    "    print(g_prok_df.columns)\n",
    "dict_groupOfAssembly     = dict(zip(g_prok_df[\"Assembly Accession\"], g_prok_df[\"Group\"]))         \n",
    "dict_subgroupOfAssembly  = dict(zip(g_prok_df[\"Assembly Accession\"], g_prok_df[\"SubGroup\"]))    \n",
    "dict_taxidOfAssembly     = dict(zip(g_prok_df[\"Assembly Accession\"], g_prok_df[\"TaxID\"]))            # print(dict_taxidOfAssembly)\n",
    "dict_gcOfAssembly        = dict(zip(g_prok_df[\"Assembly Accession\"], g_prok_df[\"GC%\"]))            \n",
    "dict_sizeOfAssembly      = dict(zip(g_prok_df[\"Assembly Accession\"], g_prok_df[\"Size (Mb)\"]))       \n",
    "dict_repliconsOfAssembly = dict(zip(g_prok_df[\"Assembly Accession\"], g_prok_df[\"Replicons\"]))     \n",
    "dict_statusOfAssembly    = dict(zip(g_prok_df[\"Assembly Accession\"], g_prok_df[\"Status\"]))        \n",
    "if 1:\n",
    "  print(len(dict_sizeOfAssembly), \"Prokaryotes\")\n",
    "\n",
    "# Final dictionaries for Eukaryotes and Prokaryotes\n",
    "dict_groupOfAssembly.update(    dict_euk_groupOfAssembly)\n",
    "dict_subgroupOfAssembly.update( dict_euk_subgroupOfAssembly)\n",
    "dict_taxidOfAssembly.update(    dict_euk_taxidOfAssembly)\n",
    "dict_gcOfAssembly.update(       dict_euk_gcOfAssembly)\n",
    "dict_sizeOfAssembly.update(     dict_euk_sizeOfAssembly)\n",
    "dict_repliconsOfAssembly.update(dict_euk_repliconsOfAssembly)\n",
    "dict_statusOfAssembly.update(   dict_euk_statusOfAssembly)\n",
    "if 1:\n",
    "  print(len(dict_sizeOfAssembly), \"Eukaryotes & Prokaryotes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c64bf704-3bad-43fb-9ec4-badb1551834f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e62c4384-6b29-4153-94ac-0e730f4d02a2",
   "metadata": {},
   "source": [
    "#### Taxid/assemblies for high quality genome assemblies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6ece19c4-da30-4c12-91df-92ac2959649a",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxid_highqu_genome_l = list(dict_taxidOfAssembly.values())\n",
    "assembly_highqu_genome_l = list(dict_taxidOfAssembly.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b6c6b37-c579-4b9f-97a6-67a28036c9ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1c84b60f-5c41-401a-96fa-821a0612b257",
   "metadata": {},
   "source": [
    "### From Ensembl Tax_id file \n",
    "dict: Assembly_accession of species (species is a perfect id in this case)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bd26fcfe-7e87-46ec-be82-7a372e403018",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/emuro/Nubya/results/geneLength/outputInputFiles/some_tables/species_Ensembl_taxid/species_Ensembl_EMv2.0.tsv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>division_7</th>\n",
       "      <th>division_8</th>\n",
       "      <th>species</th>\n",
       "      <th>assembly</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>ensembl_assembly</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>archaea</td>\n",
       "      <td>methanobacterium_bryantii_gca_002287175</td>\n",
       "      <td>ASM228717v1</td>\n",
       "      <td>2161</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "      <td>GCA_002287175.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>archaea</td>\n",
       "      <td>methanobacterium_formicicum_gca_000762265</td>\n",
       "      <td>ASM76226v1</td>\n",
       "      <td>2162</td>\n",
       "      <td>Archaea; Euryarchaeota; Methanomada group; Met...</td>\n",
       "      <td>GCA_000762265.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  division_7 division_8                                    species  \\\n",
       "0   bacteria    archaea    methanobacterium_bryantii_gca_002287175   \n",
       "1   bacteria    archaea  methanobacterium_formicicum_gca_000762265   \n",
       "\n",
       "      assembly  taxonomy_id  \\\n",
       "0  ASM228717v1         2161   \n",
       "1   ASM76226v1         2162   \n",
       "\n",
       "                                             Lineage ensembl_assembly  \n",
       "0  Archaea; Euryarchaeota; Methanomada group; Met...  GCA_002287175.1  \n",
       "1  Archaea; Euryarchaeota; Methanomada group; Met...  GCA_000762265.1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(33627, 7)\n"
     ]
    }
   ],
   "source": [
    "# ENSEMBL assembly accession\n",
    "# system\n",
    "import os\n",
    "system = list(os.uname())[0]\n",
    "if system == 'Linux':\n",
    "    base_dir = \"/media/emuro/Nubya/\"\n",
    "elif system == 'Darwin':\n",
    "    base_dir = \"/Volumes/Nubya/\"\n",
    "ensembl_Id_file = base_dir + \"results/geneLength/outputInputFiles/\" + \"some_tables/species_Ensembl_taxid/\" + \"species_Ensembl_EMv2.0.tsv\" # with covid\n",
    "print(ensembl_Id_file)\n",
    "\n",
    "# retrieve data\n",
    "id_df = pd.read_csv(ensembl_Id_file, sep=\"\\t\")\n",
    "id_df = id_df[[\"species\", \"assembly_accession\"]]\n",
    "\n",
    "dict_assemblyOfSpecies = dict(zip(id_df[\"species\"], id_df[\"assembly_accession\"])) # print(len(dict_assemblyOfSpecies))\n",
    "# visualize data\n",
    "if 0:\n",
    "    print(dict_assemblyOfspecies)\n",
    "pd.set_option('display.max_columns', None)\n",
    "if 0:\n",
    "    display(id_df.head(2))\n",
    "    print(id_df.shape)\n",
    "\n",
    "# Add the assembly of the species\n",
    "stat_df[\"ensembl_assembly\"] = stat_df[\"species\"].map(dict_assemblyOfSpecies)\n",
    "if 1:\n",
    "    display(stat_df.head(2))\n",
    "    print(stat_df.shape)       "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4b41fe0-608f-4e4f-b8e5-ad56c6048ed5",
   "metadata": {},
   "source": [
    "#### Taxid and assembly accesssion with high quality genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fd98b0f6-0553-4104-86f9-c8b46e648d4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32526, 7)\n",
      "(31736, 7)\n"
     ]
    }
   ],
   "source": [
    "# taxid with high quality genome assemblies\n",
    "taxid_highqu_genome_l    = list(dict_taxidOfAssembly.values()) # taxids\n",
    "assembly_highqu_genome_l = list(dict_taxidOfAssembly.keys())   # assemblies\n",
    "\n",
    "stat_df = stat_df[stat_df[\"taxonomy_id\"].isin(taxid_highqu_genome_l)] # reduce to hquality taxonomies...\n",
    "print(stat_df.shape) # much more redundancy, the assembly commands\n",
    "#\n",
    "stat_df = stat_df[stat_df[\"ensembl_assembly\"].isin( assembly_highqu_genome_l )]\n",
    "print(stat_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5374d50-ed27-4c2d-8d8a-9fdb512c0f13",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4ec9193c-a5e5-41d2-b5a0-66cc535993e0",
   "metadata": {},
   "source": [
    "#### For each filtered species add: genome size, gc content, chromosomes,... "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0ec4d243-5ccb-4c02-9c7f-7fb833f21251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>division_7</th>\n",
       "      <th>division_8</th>\n",
       "      <th>species</th>\n",
       "      <th>assembly</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>ensembl_assembly</th>\n",
       "      <th>group</th>\n",
       "      <th>subgroup</th>\n",
       "      <th>gc_percent</th>\n",
       "      <th>size_Mbp</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>chromosomes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33266</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>dromaius_novaehollandiae</td>\n",
       "      <td>droNov1</td>\n",
       "      <td>8790</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_003342905.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>41.7</td>\n",
       "      <td>1192.250</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>mitochondrion MT:NC_002784.1/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33267</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>apteryx_haastii</td>\n",
       "      <td>aptHaa1</td>\n",
       "      <td>8823</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_003342985.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>40.8</td>\n",
       "      <td>1221.440</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33268</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>apteryx_owenii</td>\n",
       "      <td>aptOwe1</td>\n",
       "      <td>8824</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_003342965.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>40.9</td>\n",
       "      <td>1231.280</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33269</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>anas_platyrhynchos_platyrhynchos</td>\n",
       "      <td>CAU_duck1.0</td>\n",
       "      <td>8840</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_002743455.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>41.7073</td>\n",
       "      <td>1136.420</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>chromosome 1:CM008538.1; chromosome 2:CM008539...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33270</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>numida_meleagris</td>\n",
       "      <td>NumMel1.0</td>\n",
       "      <td>8996</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_002078875.2</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>41.8975</td>\n",
       "      <td>1043.260</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>chromosome 1:NC_034409.1/CM007814.2; chromosom...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33271</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>gallus_gallus</td>\n",
       "      <td>GRCg6a</td>\n",
       "      <td>9031</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_000002315.5</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>51.8</td>\n",
       "      <td>1065.370</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33273</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>serinus_canaria</td>\n",
       "      <td>SCA1</td>\n",
       "      <td>9135</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_000534875.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>42.6</td>\n",
       "      <td>1152.100</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33356</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>melopsittacus_undulatus</td>\n",
       "      <td>Melopsittacus_undulatus_6.3</td>\n",
       "      <td>13146</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_000238935.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>41.4</td>\n",
       "      <td>1117.370</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33364</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>nothoprocta_perdicaria</td>\n",
       "      <td>notPer1</td>\n",
       "      <td>30464</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_003342845.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>42.1</td>\n",
       "      <td>965.904</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33384</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>junco_hyemalis</td>\n",
       "      <td>ASM382977v1</td>\n",
       "      <td>40217</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_003829775.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>40.7</td>\n",
       "      <td>958.315</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33391</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>zonotrichia_albicollis</td>\n",
       "      <td>Zonotrichia_albicollis-1.0.1</td>\n",
       "      <td>44394</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_000385455.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>41.8</td>\n",
       "      <td>1052.600</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33427</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>anser_brachyrhynchus</td>\n",
       "      <td>ASM259213v1</td>\n",
       "      <td>132585</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_002592135.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>41.1</td>\n",
       "      <td>1116.990</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33430</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>cyanistes_caeruleus</td>\n",
       "      <td>cyaCae2</td>\n",
       "      <td>156563</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_002901205.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>41.8</td>\n",
       "      <td>1186.980</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33433</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>calidris_pugnax</td>\n",
       "      <td>ASM143184v1</td>\n",
       "      <td>198806</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_001431845.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>42.7</td>\n",
       "      <td>1229.090</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33440</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>lonchura_striata_domestica</td>\n",
       "      <td>LonStrDom1</td>\n",
       "      <td>299123</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_002197715.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>40.4</td>\n",
       "      <td>1060.170</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33443</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>apteryx_rowi</td>\n",
       "      <td>aptRow1</td>\n",
       "      <td>308060</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_003343035.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>41.6</td>\n",
       "      <td>1228.900</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33444</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>lepidothrix_coronata</td>\n",
       "      <td>Lepidothrix_coronata-1.0</td>\n",
       "      <td>321398</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_001604755.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>41.9</td>\n",
       "      <td>1079.580</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33449</th>\n",
       "      <td>vertebrates</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>calidris_pygmaea</td>\n",
       "      <td>ASM369795v1</td>\n",
       "      <td>425635</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>GCA_003697955.1</td>\n",
       "      <td>Animals</td>\n",
       "      <td>Birds</td>\n",
       "      <td>44.2</td>\n",
       "      <td>1178.370</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        division_7   division_8                           species  \\\n",
       "33266  vertebrates  vertebrates          dromaius_novaehollandiae   \n",
       "33267  vertebrates  vertebrates                   apteryx_haastii   \n",
       "33268  vertebrates  vertebrates                    apteryx_owenii   \n",
       "33269  vertebrates  vertebrates  anas_platyrhynchos_platyrhynchos   \n",
       "33270  vertebrates  vertebrates                  numida_meleagris   \n",
       "33271  vertebrates  vertebrates                     gallus_gallus   \n",
       "33273  vertebrates  vertebrates                   serinus_canaria   \n",
       "33356  vertebrates  vertebrates           melopsittacus_undulatus   \n",
       "33364  vertebrates  vertebrates            nothoprocta_perdicaria   \n",
       "33384  vertebrates  vertebrates                    junco_hyemalis   \n",
       "33391  vertebrates  vertebrates            zonotrichia_albicollis   \n",
       "33427  vertebrates  vertebrates              anser_brachyrhynchus   \n",
       "33430  vertebrates  vertebrates               cyanistes_caeruleus   \n",
       "33433  vertebrates  vertebrates                   calidris_pugnax   \n",
       "33440  vertebrates  vertebrates        lonchura_striata_domestica   \n",
       "33443  vertebrates  vertebrates                      apteryx_rowi   \n",
       "33444  vertebrates  vertebrates              lepidothrix_coronata   \n",
       "33449  vertebrates  vertebrates                  calidris_pygmaea   \n",
       "\n",
       "                           assembly  taxonomy_id  \\\n",
       "33266                       droNov1         8790   \n",
       "33267                       aptHaa1         8823   \n",
       "33268                       aptOwe1         8824   \n",
       "33269                   CAU_duck1.0         8840   \n",
       "33270                     NumMel1.0         8996   \n",
       "33271                        GRCg6a         9031   \n",
       "33273                          SCA1         9135   \n",
       "33356   Melopsittacus_undulatus_6.3        13146   \n",
       "33364                       notPer1        30464   \n",
       "33384                   ASM382977v1        40217   \n",
       "33391  Zonotrichia_albicollis-1.0.1        44394   \n",
       "33427                   ASM259213v1       132585   \n",
       "33430                       cyaCae2       156563   \n",
       "33433                   ASM143184v1       198806   \n",
       "33440                    LonStrDom1       299123   \n",
       "33443                       aptRow1       308060   \n",
       "33444      Lepidothrix_coronata-1.0       321398   \n",
       "33449                   ASM369795v1       425635   \n",
       "\n",
       "                                                 Lineage ensembl_assembly  \\\n",
       "33266  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_003342905.1   \n",
       "33267  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_003342985.1   \n",
       "33268  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_003342965.1   \n",
       "33269  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_002743455.1   \n",
       "33270  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_002078875.2   \n",
       "33271  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_000002315.5   \n",
       "33273  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_000534875.1   \n",
       "33356  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_000238935.1   \n",
       "33364  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_003342845.1   \n",
       "33384  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_003829775.1   \n",
       "33391  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_000385455.1   \n",
       "33427  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_002592135.1   \n",
       "33430  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_002901205.1   \n",
       "33433  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_001431845.1   \n",
       "33440  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_002197715.1   \n",
       "33443  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_003343035.1   \n",
       "33444  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_001604755.1   \n",
       "33449  Eukaryota; Metazoa; Chordata; Craniata; Verteb...  GCA_003697955.1   \n",
       "\n",
       "         group subgroup gc_percent  size_Mbp assembly_status  \\\n",
       "33266  Animals    Birds       41.7  1192.250        Scaffold   \n",
       "33267  Animals    Birds       40.8  1221.440        Scaffold   \n",
       "33268  Animals    Birds       40.9  1231.280        Scaffold   \n",
       "33269  Animals    Birds    41.7073  1136.420      Chromosome   \n",
       "33270  Animals    Birds    41.8975  1043.260      Chromosome   \n",
       "33271  Animals    Birds       51.8  1065.370      Chromosome   \n",
       "33273  Animals    Birds       42.6  1152.100        Scaffold   \n",
       "33356  Animals    Birds       41.4  1117.370        Scaffold   \n",
       "33364  Animals    Birds       42.1   965.904        Scaffold   \n",
       "33384  Animals    Birds       40.7   958.315        Scaffold   \n",
       "33391  Animals    Birds       41.8  1052.600        Scaffold   \n",
       "33427  Animals    Birds       41.1  1116.990        Scaffold   \n",
       "33430  Animals    Birds       41.8  1186.980        Scaffold   \n",
       "33433  Animals    Birds       42.7  1229.090        Scaffold   \n",
       "33440  Animals    Birds       40.4  1060.170        Scaffold   \n",
       "33443  Animals    Birds       41.6  1228.900        Scaffold   \n",
       "33444  Animals    Birds       41.9  1079.580        Scaffold   \n",
       "33449  Animals    Birds       44.2  1178.370        Scaffold   \n",
       "\n",
       "                                             chromosomes  \n",
       "33266                      mitochondrion MT:NC_002784.1/  \n",
       "33267                                                  -  \n",
       "33268                                                  -  \n",
       "33269  chromosome 1:CM008538.1; chromosome 2:CM008539...  \n",
       "33270  chromosome 1:NC_034409.1/CM007814.2; chromosom...  \n",
       "33271                                                  -  \n",
       "33273                                                  -  \n",
       "33356                                                  -  \n",
       "33364                                                  -  \n",
       "33384                                                  -  \n",
       "33391                                                  -  \n",
       "33427                                                  -  \n",
       "33430                                                  -  \n",
       "33433                                                  -  \n",
       "33440                                                  -  \n",
       "33443                                                  -  \n",
       "33444                                                  -  \n",
       "33449                                                  -  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18, 13)\n",
      "division_8\n",
      "vertebrates    18\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Add the assembly of the species\n",
    "stat_df[\"ensembl_assembly\"] = stat_df[\"species\"].map(dict_assemblyOfSpecies)\n",
    "# Add more genomic info\n",
    "stat_df[\"group\"]      = stat_df[\"ensembl_assembly\"].map(dict_groupOfAssembly)\n",
    "stat_df[\"subgroup\"]        = stat_df[\"ensembl_assembly\"].map(dict_subgroupOfAssembly)\n",
    "stat_df[\"gc_percent\"]      = stat_df[\"ensembl_assembly\"].map(dict_gcOfAssembly)\n",
    "stat_df[\"size_Mbp\"]        = stat_df[\"ensembl_assembly\"].map(dict_sizeOfAssembly)\n",
    "stat_df[\"assembly_status\"] = stat_df[\"ensembl_assembly\"].map(dict_statusOfAssembly)\n",
    "stat_df[\"chromosomes\"]     = stat_df[\"ensembl_assembly\"].map(dict_repliconsOfAssembly)\n",
    "\n",
    "stat_df = stat_df[stat_df.division_8 != \"bacteria\"]\n",
    "stat_df = stat_df[stat_df.division_8 != \"archaea\"]\n",
    "#stat_df = stat_df[stat_df.division_8 != \"fungi\"]\n",
    "#stat_df = stat_df[stat_df.division_8 != \"plants\"]\n",
    "#stat_df = stat_df[stat_df.division_8 != \"protists\"]\n",
    "#stat_df = stat_df[stat_df.division_8 != \"archaea\"]\n",
    "stat_df = stat_df[stat_df.subgroup == \"Birds\"]\n",
    "#stat_df = stat_df[stat_df.subgroup == \"Fishes\"]\n",
    "\n",
    "display(stat_df.head(40))\n",
    "print(stat_df.shape)\n",
    "print(stat_df['division_8'].value_counts(dropna=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa8f75b4-cdea-4d26-83e4-e8c017ce4dd1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcd14357-9cfa-41ca-bb6e-097985d55c6c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
