{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aeef9d12-d367-4aa1-bd3a-261c285248b4",
   "metadata": {},
   "source": [
    "# Genes' bias per chr\n",
    "\n",
    "Repito lo que hice hace a√±os (cuaderno II. 17 Nov 2016), para ver si hay sesgo en el strand de los genes en los chrs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fbe456d-5e46-44b7-bf82-4d748cb291fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "51b0a90c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy import stats\n",
    "import os \n",
    "import pandas as pd\n",
    "from plotnine import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23d40525-65aa-4a58-881c-72a93960fcbb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cd709b0c-db39-4678-9a34-fc82a6ad07dd",
   "metadata": {},
   "source": [
    "## Observations: Protein length distribution in different species"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48adde40-48ca-45cb-9cee-cfa11cfab6ff",
   "metadata": {},
   "source": [
    "### Retrieve the statistical description and tax_id of species"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39125c32-b17e-44f7-b580-9ba326045f0f",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Statistical description data file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d892fcdf-a696-4cc1-9cd8-0f888c093351",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Wes/results/geneLength/outputInputFiles/analysis/some_statistics/stat_description/taxid_merged/stat_description.taxid_merged.ensembl_and_ref_proteome.tsv\n",
      "(6521, 66)\n"
     ]
    }
   ],
   "source": [
    "# system\n",
    "system = list(os.uname())[0]\n",
    "if system == 'Linux':\n",
    "    base_dir = \"/media/emuro/Wes/\"\n",
    "elif system == 'Darwin':\n",
    "    base_dir = \"/Volumes/Wes/\"\n",
    "    \n",
    "# statistics on length distribution for different species     \n",
    "stat_file = base_dir + \"results/geneLength/outputInputFiles/\" + \"analysis/some_statistics/stat_description/\" \n",
    "stat_file += \"taxid_merged/stat_description.taxid_merged.ensembl_and_ref_proteome.tsv\"\n",
    "print(stat_file)\n",
    "\n",
    "# retrieve data\n",
    "stat_df = pd.read_csv(stat_file, low_memory=False, sep=\"\\t\")\n",
    "print(stat_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7ff1aed-e92e-4b2a-82c3-c6bed069e33f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9964c1d5-b4e1-4998-a819-f93af00642aa",
   "metadata": {},
   "source": [
    "##### Filter some columns and species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "90d2b7e7-99db-441c-b7d6-f6509e017fa7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes_species</th>\n",
       "      <th>genes_assembly</th>\n",
       "      <th>genes_trunk_genes_path</th>\n",
       "      <th>genes_genes_file</th>\n",
       "      <th>genes_count</th>\n",
       "      <th>genes_mean</th>\n",
       "      <th>genes_std</th>\n",
       "      <th>genes_var</th>\n",
       "      <th>genes_min</th>\n",
       "      <th>genes_25perc</th>\n",
       "      <th>genes_50perc</th>\n",
       "      <th>genes_75perc</th>\n",
       "      <th>genes_max</th>\n",
       "      <th>genes_log10_mean</th>\n",
       "      <th>genes_log10_std</th>\n",
       "      <th>genes_log10_var</th>\n",
       "      <th>genes_log10_min</th>\n",
       "      <th>genes_log10_25perc</th>\n",
       "      <th>genes_log10_50perc</th>\n",
       "      <th>genes_log10_75perc</th>\n",
       "      <th>genes_log10_max</th>\n",
       "      <th>genes_log_mean</th>\n",
       "      <th>genes_log_std</th>\n",
       "      <th>genes_log_var</th>\n",
       "      <th>genes_log_min</th>\n",
       "      <th>genes_log_25perc</th>\n",
       "      <th>genes_log_50perc</th>\n",
       "      <th>genes_log_75perc</th>\n",
       "      <th>genes_log_max</th>\n",
       "      <th>genes_division</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>prots_species</th>\n",
       "      <th>prots_proteome_id</th>\n",
       "      <th>prots_superregnum</th>\n",
       "      <th>prots_num_prot_cod_genes</th>\n",
       "      <th>prots_uniprot_fasta_file</th>\n",
       "      <th>prots_count</th>\n",
       "      <th>prots_mean</th>\n",
       "      <th>prots_std</th>\n",
       "      <th>prots_var</th>\n",
       "      <th>prots_min</th>\n",
       "      <th>prots_25perc</th>\n",
       "      <th>prots_50perc</th>\n",
       "      <th>prots_75perc</th>\n",
       "      <th>prots_max</th>\n",
       "      <th>prots_log10_mean</th>\n",
       "      <th>prots_log10_std</th>\n",
       "      <th>prots_log10_var</th>\n",
       "      <th>prots_log10_min</th>\n",
       "      <th>prots_log10_25perc</th>\n",
       "      <th>prots_log10_50perc</th>\n",
       "      <th>prots_log10_75perc</th>\n",
       "      <th>prots_log10_max</th>\n",
       "      <th>prots_log_mean</th>\n",
       "      <th>prots_log_std</th>\n",
       "      <th>prots_log_var</th>\n",
       "      <th>prots_log_min</th>\n",
       "      <th>prots_log_25perc</th>\n",
       "      <th>prots_log_50perc</th>\n",
       "      <th>prots_log_75perc</th>\n",
       "      <th>prots_log_max</th>\n",
       "      <th>division_8</th>\n",
       "      <th>ratio_mean_prots_genes</th>\n",
       "      <th>diff_prots_genes</th>\n",
       "      <th>abs_diff</th>\n",
       "      <th>ratio_prots_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>archangium_gephyra_gca_001027285</td>\n",
       "      <td>ASM102728v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.archangium_gephyra_gca_00...</td>\n",
       "      <td>10119.0</td>\n",
       "      <td>1103.871035</td>\n",
       "      <td>1127.066255</td>\n",
       "      <td>1.270278e+06</td>\n",
       "      <td>113.0</td>\n",
       "      <td>524.0</td>\n",
       "      <td>887.0</td>\n",
       "      <td>1388.0</td>\n",
       "      <td>36653.0</td>\n",
       "      <td>2.919254</td>\n",
       "      <td>0.333869</td>\n",
       "      <td>0.111468</td>\n",
       "      <td>2.053078</td>\n",
       "      <td>2.719331</td>\n",
       "      <td>2.947924</td>\n",
       "      <td>3.142389</td>\n",
       "      <td>4.56411</td>\n",
       "      <td>6.721831</td>\n",
       "      <td>0.768762</td>\n",
       "      <td>0.590995</td>\n",
       "      <td>4.727388</td>\n",
       "      <td>6.261492</td>\n",
       "      <td>6.787845</td>\n",
       "      <td>7.235619</td>\n",
       "      <td>10.509251</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>48</td>\n",
       "      <td>Archangium gephyra</td>\n",
       "      <td>UP000035579</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>10110</td>\n",
       "      <td>/ftp.uniprot.org/pub/databases/uniprot/current...</td>\n",
       "      <td>10110.0</td>\n",
       "      <td>367.499703</td>\n",
       "      <td>375.772244</td>\n",
       "      <td>141204.779058</td>\n",
       "      <td>37.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>295.0</td>\n",
       "      <td>462.0</td>\n",
       "      <td>12217.0</td>\n",
       "      <td>2.441094</td>\n",
       "      <td>0.334836</td>\n",
       "      <td>0.112115</td>\n",
       "      <td>1.568202</td>\n",
       "      <td>2.240549</td>\n",
       "      <td>2.469822</td>\n",
       "      <td>2.664642</td>\n",
       "      <td>4.086965</td>\n",
       "      <td>5.620827</td>\n",
       "      <td>0.770987</td>\n",
       "      <td>0.594422</td>\n",
       "      <td>3.610918</td>\n",
       "      <td>5.159055</td>\n",
       "      <td>5.686975</td>\n",
       "      <td>6.135565</td>\n",
       "      <td>9.410584</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>0.998757</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.999111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chondromyces_crocatus_gca_001189295</td>\n",
       "      <td>ASM118929v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.chondromyces_crocatus_gca...</td>\n",
       "      <td>8339.0</td>\n",
       "      <td>1211.514450</td>\n",
       "      <td>1051.408605</td>\n",
       "      <td>1.105460e+06</td>\n",
       "      <td>89.0</td>\n",
       "      <td>629.0</td>\n",
       "      <td>977.0</td>\n",
       "      <td>1460.0</td>\n",
       "      <td>24476.0</td>\n",
       "      <td>2.978963</td>\n",
       "      <td>0.300250</td>\n",
       "      <td>0.090150</td>\n",
       "      <td>1.949390</td>\n",
       "      <td>2.798651</td>\n",
       "      <td>2.989895</td>\n",
       "      <td>3.164353</td>\n",
       "      <td>4.38874</td>\n",
       "      <td>6.859315</td>\n",
       "      <td>0.691352</td>\n",
       "      <td>0.477968</td>\n",
       "      <td>4.488636</td>\n",
       "      <td>6.444131</td>\n",
       "      <td>6.884487</td>\n",
       "      <td>7.286192</td>\n",
       "      <td>10.105448</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>52</td>\n",
       "      <td>Chondromyces crocatus</td>\n",
       "      <td>UP000067626</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>8327</td>\n",
       "      <td>/ftp.uniprot.org/pub/databases/uniprot/current...</td>\n",
       "      <td>8327.0</td>\n",
       "      <td>403.466555</td>\n",
       "      <td>350.579560</td>\n",
       "      <td>122906.027917</td>\n",
       "      <td>29.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>486.0</td>\n",
       "      <td>8158.0</td>\n",
       "      <td>2.501095</td>\n",
       "      <td>0.300984</td>\n",
       "      <td>0.090591</td>\n",
       "      <td>1.462398</td>\n",
       "      <td>2.320146</td>\n",
       "      <td>2.513218</td>\n",
       "      <td>2.686636</td>\n",
       "      <td>3.911584</td>\n",
       "      <td>5.758984</td>\n",
       "      <td>0.693041</td>\n",
       "      <td>0.480306</td>\n",
       "      <td>3.367296</td>\n",
       "      <td>5.342334</td>\n",
       "      <td>5.786897</td>\n",
       "      <td>6.186209</td>\n",
       "      <td>9.006754</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>0.999080</td>\n",
       "      <td>-12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.998561</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         genes_species genes_assembly  \\\n",
       "0     archangium_gephyra_gca_001027285    ASM102728v1   \n",
       "1  chondromyces_crocatus_gca_001189295    ASM118929v1   \n",
       "\n",
       "                              genes_trunk_genes_path  \\\n",
       "0  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "1  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "\n",
       "                                    genes_genes_file  genes_count  \\\n",
       "0  protein_coding.genes.archangium_gephyra_gca_00...      10119.0   \n",
       "1  protein_coding.genes.chondromyces_crocatus_gca...       8339.0   \n",
       "\n",
       "    genes_mean    genes_std     genes_var  genes_min  genes_25perc  \\\n",
       "0  1103.871035  1127.066255  1.270278e+06      113.0         524.0   \n",
       "1  1211.514450  1051.408605  1.105460e+06       89.0         629.0   \n",
       "\n",
       "   genes_50perc  genes_75perc  genes_max  genes_log10_mean  genes_log10_std  \\\n",
       "0         887.0        1388.0    36653.0          2.919254         0.333869   \n",
       "1         977.0        1460.0    24476.0          2.978963         0.300250   \n",
       "\n",
       "   genes_log10_var  genes_log10_min  genes_log10_25perc  genes_log10_50perc  \\\n",
       "0         0.111468         2.053078            2.719331            2.947924   \n",
       "1         0.090150         1.949390            2.798651            2.989895   \n",
       "\n",
       "   genes_log10_75perc  genes_log10_max  genes_log_mean  genes_log_std  \\\n",
       "0            3.142389          4.56411        6.721831       0.768762   \n",
       "1            3.164353          4.38874        6.859315       0.691352   \n",
       "\n",
       "   genes_log_var  genes_log_min  genes_log_25perc  genes_log_50perc  \\\n",
       "0       0.590995       4.727388          6.261492          6.787845   \n",
       "1       0.477968       4.488636          6.444131          6.884487   \n",
       "\n",
       "   genes_log_75perc  genes_log_max genes_division  tax_id  \\\n",
       "0          7.235619      10.509251       bacteria      48   \n",
       "1          7.286192      10.105448       bacteria      52   \n",
       "\n",
       "           prots_species prots_proteome_id prots_superregnum  \\\n",
       "0     Archangium gephyra       UP000035579          bacteria   \n",
       "1  Chondromyces crocatus       UP000067626          bacteria   \n",
       "\n",
       "   prots_num_prot_cod_genes  \\\n",
       "0                     10110   \n",
       "1                      8327   \n",
       "\n",
       "                            prots_uniprot_fasta_file  prots_count  prots_mean  \\\n",
       "0  /ftp.uniprot.org/pub/databases/uniprot/current...      10110.0  367.499703   \n",
       "1  /ftp.uniprot.org/pub/databases/uniprot/current...       8327.0  403.466555   \n",
       "\n",
       "    prots_std      prots_var  prots_min  prots_25perc  prots_50perc  \\\n",
       "0  375.772244  141204.779058       37.0         174.0         295.0   \n",
       "1  350.579560  122906.027917       29.0         209.0         326.0   \n",
       "\n",
       "   prots_75perc  prots_max  prots_log10_mean  prots_log10_std  \\\n",
       "0         462.0    12217.0          2.441094         0.334836   \n",
       "1         486.0     8158.0          2.501095         0.300984   \n",
       "\n",
       "   prots_log10_var  prots_log10_min  prots_log10_25perc  prots_log10_50perc  \\\n",
       "0         0.112115         1.568202            2.240549            2.469822   \n",
       "1         0.090591         1.462398            2.320146            2.513218   \n",
       "\n",
       "   prots_log10_75perc  prots_log10_max  prots_log_mean  prots_log_std  \\\n",
       "0            2.664642         4.086965        5.620827       0.770987   \n",
       "1            2.686636         3.911584        5.758984       0.693041   \n",
       "\n",
       "   prots_log_var  prots_log_min  prots_log_25perc  prots_log_50perc  \\\n",
       "0       0.594422       3.610918          5.159055          5.686975   \n",
       "1       0.480306       3.367296          5.342334          5.786897   \n",
       "\n",
       "   prots_log_75perc  prots_log_max division_8  ratio_mean_prots_genes  \\\n",
       "0          6.135565       9.410584   bacteria                0.998757   \n",
       "1          6.186209       9.006754   bacteria                0.999080   \n",
       "\n",
       "   diff_prots_genes  abs_diff  ratio_prots_genes  \n",
       "0              -9.0       9.0           0.999111  \n",
       "1             -12.0      12.0           0.998561  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6521, 66)\n"
     ]
    }
   ],
   "source": [
    "stat_df.rename(columns={'merged_division_superregnum': 'division_8'},\n",
    "               inplace=True)\n",
    "# visualize data\n",
    "pd.set_option('display.max_columns', None)\n",
    "if 1:\n",
    "    display(stat_df.head(2))\n",
    "    print(stat_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2029fd0-59cd-43f1-966f-91ba92732e17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "02f530e2-47f1-4cd1-85fd-7f0e16c65029",
   "metadata": {},
   "source": [
    "# Python functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af6c0ad2-9985-4285-887a-c413505d46e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# species' stats for length distribution\n",
    "def get_df_for_taxid(df, taxid):\n",
    "     return df.loc[df[\"tax_id\"]==taxid].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aff7ede5-eaad-4230-a0ea-697ca18d0f43",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The next function performs the whole analysis\n",
    "def species_gene_distribution__retrieve_plot_and_fit(stat_taxid_df, tax_id):\n",
    "    species_df = get_df_for_taxid(stat_taxid_df, tax_id)\n",
    "    species_df.drop_duplicates(subset=['tax_id'], inplace=True, keep='last')\n",
    "    display(species_df)\n",
    "    \n",
    "    genes_f = base_dir + \"results/geneLength/\" + species_df[\"genes_trunk_genes_path\"].item() + species_df[\"genes_genes_file\"].item()\n",
    "    print(genes_f)\n",
    "    # retrieve data\n",
    "    genes_df = pd.read_csv(genes_f, sep=\"\\t\")\n",
    "    genes_df[\"TSS\"] = np.where(genes_df['strand']!= '+', genes_df['start'], genes_df['end'])\n",
    "    \n",
    "    # visualize data\n",
    "    pd.set_option('display.max_columns', None)\n",
    "    if 1:\n",
    "        display(genes_df.head(20))\n",
    "        print(genes_df.shape)\n",
    "\n",
    "    return(genes_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67ede334-5d2f-4d20-86d6-887eb53c4a3d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d0dd8ca3-6d99-4edf-bb59-03d4187d8385",
   "metadata": {},
   "source": [
    "# Calculate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b468cfd2-90f5-47e0-a96f-48bc7c39a11b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes_species</th>\n",
       "      <th>genes_assembly</th>\n",
       "      <th>genes_trunk_genes_path</th>\n",
       "      <th>genes_genes_file</th>\n",
       "      <th>genes_count</th>\n",
       "      <th>genes_mean</th>\n",
       "      <th>genes_std</th>\n",
       "      <th>genes_var</th>\n",
       "      <th>genes_min</th>\n",
       "      <th>genes_25perc</th>\n",
       "      <th>genes_50perc</th>\n",
       "      <th>genes_75perc</th>\n",
       "      <th>genes_max</th>\n",
       "      <th>genes_log10_mean</th>\n",
       "      <th>genes_log10_std</th>\n",
       "      <th>genes_log10_var</th>\n",
       "      <th>genes_log10_min</th>\n",
       "      <th>genes_log10_25perc</th>\n",
       "      <th>genes_log10_50perc</th>\n",
       "      <th>genes_log10_75perc</th>\n",
       "      <th>genes_log10_max</th>\n",
       "      <th>genes_log_mean</th>\n",
       "      <th>genes_log_std</th>\n",
       "      <th>genes_log_var</th>\n",
       "      <th>genes_log_min</th>\n",
       "      <th>genes_log_25perc</th>\n",
       "      <th>genes_log_50perc</th>\n",
       "      <th>genes_log_75perc</th>\n",
       "      <th>genes_log_max</th>\n",
       "      <th>genes_division</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>prots_species</th>\n",
       "      <th>prots_proteome_id</th>\n",
       "      <th>prots_superregnum</th>\n",
       "      <th>prots_num_prot_cod_genes</th>\n",
       "      <th>prots_uniprot_fasta_file</th>\n",
       "      <th>prots_count</th>\n",
       "      <th>prots_mean</th>\n",
       "      <th>prots_std</th>\n",
       "      <th>prots_var</th>\n",
       "      <th>prots_min</th>\n",
       "      <th>prots_25perc</th>\n",
       "      <th>prots_50perc</th>\n",
       "      <th>prots_75perc</th>\n",
       "      <th>prots_max</th>\n",
       "      <th>prots_log10_mean</th>\n",
       "      <th>prots_log10_std</th>\n",
       "      <th>prots_log10_var</th>\n",
       "      <th>prots_log10_min</th>\n",
       "      <th>prots_log10_25perc</th>\n",
       "      <th>prots_log10_50perc</th>\n",
       "      <th>prots_log10_75perc</th>\n",
       "      <th>prots_log10_max</th>\n",
       "      <th>prots_log_mean</th>\n",
       "      <th>prots_log_std</th>\n",
       "      <th>prots_log_var</th>\n",
       "      <th>prots_log_min</th>\n",
       "      <th>prots_log_25perc</th>\n",
       "      <th>prots_log_50perc</th>\n",
       "      <th>prots_log_75perc</th>\n",
       "      <th>prots_log_max</th>\n",
       "      <th>division_8</th>\n",
       "      <th>ratio_mean_prots_genes</th>\n",
       "      <th>diff_prots_genes</th>\n",
       "      <th>abs_diff</th>\n",
       "      <th>ratio_prots_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ftp.ensembl.org/pub/release-98/genes/homo_sapi...</td>\n",
       "      <td>protein_coding.genes.homo_sapiens.nan.ensembl....</td>\n",
       "      <td>19976.0</td>\n",
       "      <td>68287.089808</td>\n",
       "      <td>132423.575448</td>\n",
       "      <td>1.753600e+10</td>\n",
       "      <td>116.0</td>\n",
       "      <td>9606.75</td>\n",
       "      <td>27221.0</td>\n",
       "      <td>70804.0</td>\n",
       "      <td>2473538.0</td>\n",
       "      <td>4.40396</td>\n",
       "      <td>0.644353</td>\n",
       "      <td>0.415191</td>\n",
       "      <td>2.064458</td>\n",
       "      <td>3.982576</td>\n",
       "      <td>4.434904</td>\n",
       "      <td>4.850058</td>\n",
       "      <td>6.393319</td>\n",
       "      <td>10.140494</td>\n",
       "      <td>1.483677</td>\n",
       "      <td>2.201298</td>\n",
       "      <td>4.75359</td>\n",
       "      <td>9.170221</td>\n",
       "      <td>10.211744</td>\n",
       "      <td>11.167671</td>\n",
       "      <td>14.72116</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>UP000005640</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>20614</td>\n",
       "      <td>/ftp.uniprot.org/pub/databases/uniprot/current...</td>\n",
       "      <td>20614.0</td>\n",
       "      <td>552.845105</td>\n",
       "      <td>595.745883</td>\n",
       "      <td>354913.157154</td>\n",
       "      <td>2.0</td>\n",
       "      <td>245.0</td>\n",
       "      <td>410.0</td>\n",
       "      <td>666.0</td>\n",
       "      <td>34350.0</td>\n",
       "      <td>2.601146</td>\n",
       "      <td>0.353891</td>\n",
       "      <td>0.125239</td>\n",
       "      <td>0.30103</td>\n",
       "      <td>2.389166</td>\n",
       "      <td>2.612784</td>\n",
       "      <td>2.823474</td>\n",
       "      <td>4.535927</td>\n",
       "      <td>5.989361</td>\n",
       "      <td>0.814863</td>\n",
       "      <td>0.664002</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>5.501258</td>\n",
       "      <td>6.016157</td>\n",
       "      <td>6.50129</td>\n",
       "      <td>10.444357</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>0.024288</td>\n",
       "      <td>638.0</td>\n",
       "      <td>638.0</td>\n",
       "      <td>1.031938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    genes_species genes_assembly  \\\n",
       "213  homo_sapiens         GRCh38   \n",
       "\n",
       "                                genes_trunk_genes_path  \\\n",
       "213  ftp.ensembl.org/pub/release-98/genes/homo_sapi...   \n",
       "\n",
       "                                      genes_genes_file  genes_count  \\\n",
       "213  protein_coding.genes.homo_sapiens.nan.ensembl....      19976.0   \n",
       "\n",
       "       genes_mean      genes_std     genes_var  genes_min  genes_25perc  \\\n",
       "213  68287.089808  132423.575448  1.753600e+10      116.0       9606.75   \n",
       "\n",
       "     genes_50perc  genes_75perc  genes_max  genes_log10_mean  genes_log10_std  \\\n",
       "213       27221.0       70804.0  2473538.0           4.40396         0.644353   \n",
       "\n",
       "     genes_log10_var  genes_log10_min  genes_log10_25perc  genes_log10_50perc  \\\n",
       "213         0.415191         2.064458            3.982576            4.434904   \n",
       "\n",
       "     genes_log10_75perc  genes_log10_max  genes_log_mean  genes_log_std  \\\n",
       "213            4.850058         6.393319       10.140494       1.483677   \n",
       "\n",
       "     genes_log_var  genes_log_min  genes_log_25perc  genes_log_50perc  \\\n",
       "213       2.201298        4.75359          9.170221         10.211744   \n",
       "\n",
       "     genes_log_75perc  genes_log_max genes_division  tax_id  \\\n",
       "213         11.167671       14.72116    vertebrates    9606   \n",
       "\n",
       "            prots_species prots_proteome_id prots_superregnum  \\\n",
       "213  Homo sapiens (Human)       UP000005640         eukaryota   \n",
       "\n",
       "     prots_num_prot_cod_genes  \\\n",
       "213                     20614   \n",
       "\n",
       "                              prots_uniprot_fasta_file  prots_count  \\\n",
       "213  /ftp.uniprot.org/pub/databases/uniprot/current...      20614.0   \n",
       "\n",
       "     prots_mean   prots_std      prots_var  prots_min  prots_25perc  \\\n",
       "213  552.845105  595.745883  354913.157154        2.0         245.0   \n",
       "\n",
       "     prots_50perc  prots_75perc  prots_max  prots_log10_mean  prots_log10_std  \\\n",
       "213         410.0         666.0    34350.0          2.601146         0.353891   \n",
       "\n",
       "     prots_log10_var  prots_log10_min  prots_log10_25perc  prots_log10_50perc  \\\n",
       "213         0.125239          0.30103            2.389166            2.612784   \n",
       "\n",
       "     prots_log10_75perc  prots_log10_max  prots_log_mean  prots_log_std  \\\n",
       "213            2.823474         4.535927        5.989361       0.814863   \n",
       "\n",
       "     prots_log_var  prots_log_min  prots_log_25perc  prots_log_50perc  \\\n",
       "213       0.664002       0.693147          5.501258          6.016157   \n",
       "\n",
       "     prots_log_75perc  prots_log_max   division_8  ratio_mean_prots_genes  \\\n",
       "213           6.50129      10.444357  vertebrates                0.024288   \n",
       "\n",
       "     diff_prots_genes  abs_diff  ratio_prots_genes  \n",
       "213             638.0     638.0           1.031938  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Wes/results/geneLength/ftp.ensembl.org/pub/release-98/genes/homo_sapiens/protein_coding.genes.homo_sapiens.nan.ensembl.98.tsv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>biotype</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>diffLength</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>49529869</td>\n",
       "      <td>49529985</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000236737</td>\n",
       "      <td>GAGE12B</td>\n",
       "      <td>116</td>\n",
       "      <td>18</td>\n",
       "      <td>49529985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>186401046</td>\n",
       "      <td>186401180</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000262180</td>\n",
       "      <td>OCLM</td>\n",
       "      <td>134</td>\n",
       "      <td>18</td>\n",
       "      <td>186401180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>131035092</td>\n",
       "      <td>131035262</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000284479</td>\n",
       "      <td>SMIM39</td>\n",
       "      <td>170</td>\n",
       "      <td>6</td>\n",
       "      <td>131035262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>135927274</td>\n",
       "      <td>135927450</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000267697</td>\n",
       "      <td>LUZP6</td>\n",
       "      <td>176</td>\n",
       "      <td>3</td>\n",
       "      <td>135927274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>135309480</td>\n",
       "      <td>135309659</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000283644</td>\n",
       "      <td>ETDC</td>\n",
       "      <td>179</td>\n",
       "      <td>3</td>\n",
       "      <td>135309659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>+</td>\n",
       "      <td>32647202</td>\n",
       "      <td>32647390</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000286131</td>\n",
       "      <td>AC083977.1</td>\n",
       "      <td>188</td>\n",
       "      <td>3</td>\n",
       "      <td>32647390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>229305135</td>\n",
       "      <td>229305326</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000213029</td>\n",
       "      <td>SPHAR</td>\n",
       "      <td>191</td>\n",
       "      <td>0</td>\n",
       "      <td>229305326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>20</td>\n",
       "      <td>-</td>\n",
       "      <td>10420546</td>\n",
       "      <td>10420737</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000285723</td>\n",
       "      <td>AL034430.2</td>\n",
       "      <td>191</td>\n",
       "      <td>0</td>\n",
       "      <td>10420546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>46326288</td>\n",
       "      <td>46326491</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000286224</td>\n",
       "      <td>AP000471.1</td>\n",
       "      <td>203</td>\n",
       "      <td>3</td>\n",
       "      <td>46326491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>MT</td>\n",
       "      <td>+</td>\n",
       "      <td>8366</td>\n",
       "      <td>8572</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>206</td>\n",
       "      <td>3</td>\n",
       "      <td>8572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>21</td>\n",
       "      <td>-</td>\n",
       "      <td>30348399</td>\n",
       "      <td>30348609</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000186980</td>\n",
       "      <td>KRTAP23-1</td>\n",
       "      <td>210</td>\n",
       "      <td>2</td>\n",
       "      <td>30348399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>22</td>\n",
       "      <td>+</td>\n",
       "      <td>39504231</td>\n",
       "      <td>39504443</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000285025</td>\n",
       "      <td>AL022312.1</td>\n",
       "      <td>212</td>\n",
       "      <td>2</td>\n",
       "      <td>39504443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>88521573</td>\n",
       "      <td>88521789</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000255072</td>\n",
       "      <td>PIGY</td>\n",
       "      <td>216</td>\n",
       "      <td>4</td>\n",
       "      <td>88521573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>30620627</td>\n",
       "      <td>30620850</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000206105</td>\n",
       "      <td>KRTAP20-4</td>\n",
       "      <td>223</td>\n",
       "      <td>7</td>\n",
       "      <td>30620850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>21</td>\n",
       "      <td>-</td>\n",
       "      <td>30718525</td>\n",
       "      <td>30718777</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000231068</td>\n",
       "      <td>KRTAP21-3</td>\n",
       "      <td>252</td>\n",
       "      <td>5</td>\n",
       "      <td>30718525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>227666804</td>\n",
       "      <td>227667061</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000284667</td>\n",
       "      <td>SCYGR5</td>\n",
       "      <td>257</td>\n",
       "      <td>5</td>\n",
       "      <td>227667061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>227387683</td>\n",
       "      <td>227387949</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000284629</td>\n",
       "      <td>SCYGR1</td>\n",
       "      <td>266</td>\n",
       "      <td>3</td>\n",
       "      <td>227387683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>152985231</td>\n",
       "      <td>152985500</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000169474</td>\n",
       "      <td>SPRR1A</td>\n",
       "      <td>269</td>\n",
       "      <td>0</td>\n",
       "      <td>152985500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>152622834</td>\n",
       "      <td>152623103</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000185962</td>\n",
       "      <td>LCE3A</td>\n",
       "      <td>269</td>\n",
       "      <td>0</td>\n",
       "      <td>152622834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>30642864</td>\n",
       "      <td>30643136</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000206104</td>\n",
       "      <td>KRTAP20-3</td>\n",
       "      <td>272</td>\n",
       "      <td>2</td>\n",
       "      <td>30643136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0 contig strand      start        end         biotype  \\\n",
       "0            0      X      +   49529869   49529985  protein_coding   \n",
       "1            1      1      +  186401046  186401180  protein_coding   \n",
       "2            2      2      +  131035092  131035262  protein_coding   \n",
       "3            3      7      -  135927274  135927450  protein_coding   \n",
       "4            4      X      +  135309480  135309659  protein_coding   \n",
       "5            5      8      +   32647202   32647390  protein_coding   \n",
       "6            6      1      +  229305135  229305326  protein_coding   \n",
       "7            7     20      -   10420546   10420737  protein_coding   \n",
       "8            8     21      +   46326288   46326491  protein_coding   \n",
       "9            9     MT      +       8366       8572  protein_coding   \n",
       "10          10     21      -   30348399   30348609  protein_coding   \n",
       "11          11     22      +   39504231   39504443  protein_coding   \n",
       "12          12      4      -   88521573   88521789  protein_coding   \n",
       "13          13     21      +   30620627   30620850  protein_coding   \n",
       "14          14     21      -   30718525   30718777  protein_coding   \n",
       "15          15      2      +  227666804  227667061  protein_coding   \n",
       "16          16      2      -  227387683  227387949  protein_coding   \n",
       "17          17      1      +  152985231  152985500  protein_coding   \n",
       "18          18      1      -  152622834  152623103  protein_coding   \n",
       "19          19     21      +   30642864   30643136  protein_coding   \n",
       "\n",
       "            gene_id   gene_name  length  diffLength        TSS  \n",
       "0   ENSG00000236737     GAGE12B     116          18   49529985  \n",
       "1   ENSG00000262180        OCLM     134          18  186401180  \n",
       "2   ENSG00000284479      SMIM39     170           6  131035262  \n",
       "3   ENSG00000267697       LUZP6     176           3  135927274  \n",
       "4   ENSG00000283644        ETDC     179           3  135309659  \n",
       "5   ENSG00000286131  AC083977.1     188           3   32647390  \n",
       "6   ENSG00000213029       SPHAR     191           0  229305326  \n",
       "7   ENSG00000285723  AL034430.2     191           0   10420546  \n",
       "8   ENSG00000286224  AP000471.1     203           3   46326491  \n",
       "9   ENSG00000228253     MT-ATP8     206           3       8572  \n",
       "10  ENSG00000186980   KRTAP23-1     210           2   30348399  \n",
       "11  ENSG00000285025  AL022312.1     212           2   39504443  \n",
       "12  ENSG00000255072        PIGY     216           4   88521573  \n",
       "13  ENSG00000206105   KRTAP20-4     223           7   30620850  \n",
       "14  ENSG00000231068   KRTAP21-3     252           5   30718525  \n",
       "15  ENSG00000284667      SCYGR5     257           5  227667061  \n",
       "16  ENSG00000284629      SCYGR1     266           3  227387683  \n",
       "17  ENSG00000169474      SPRR1A     269           0  152985500  \n",
       "18  ENSG00000185962       LCE3A     269           0  152622834  \n",
       "19  ENSG00000206104   KRTAP20-3     272           2   30643136  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19976, 11)\n"
     ]
    }
   ],
   "source": [
    "human_tax_id = 9606\n",
    "    \n",
    "genes_df = species_gene_distribution__retrieve_plot_and_fit(stat_df, human_tax_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a548f985-2ddb-464e-b3c6-0fa738c3bad4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2c648cf9-3f13-4f7b-871a-831c76f395d7",
   "metadata": {},
   "source": [
    "## Chromosomes  \n",
    "Centromeres  \n",
    "https://www.biostars.org/p/2349/\n",
    "\n",
    "curl -s \"http://hgdownload.cse.ucsc.edu/goldenPath/hg38/database/cytoBand.txt.gz\" | gunzip -c | grep acen > hg38__acen_cytoBand.txt\n",
    "\n",
    "quick try: EM manually parses to EM_hg38__acen_cytoBand.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dfff40f9-0a6d-4d67-a879-8e300dbbad86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>cen_start</th>\n",
       "      <th>cen_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>121700000</td>\n",
       "      <td>125100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>38000000</td>\n",
       "      <td>41600000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  contig  cen_start    cen_end\n",
       "0      1  121700000  125100000\n",
       "1     10   38000000   41600000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "centromere_file = hg38_centromere_file = \"./EM_hg38__acen_cytoBand.txt\"\n",
    "# retrieve data\n",
    "cent_df = pd.read_csv(centromere_file, sep=\"\\t\", names=[\"contig\", \"cen_start\", \"cen_end\", \"arm\", \"acen\"])\n",
    "\n",
    "cent_df[\"contig\"] = cent_df[\"contig\"].str.replace(\"chr\", \"\")\n",
    "contig_l = cent_df[\"contig\"].to_list()\n",
    "\n",
    "cent_df = cent_df[[\"contig\", \"cen_start\", \"cen_end\"]]\n",
    "print(cent_df.shape)\n",
    "display(cent_df.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9ca71de8-e187-47ab-ba26-d6248b190f29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>contig</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>biotype</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>length</th>\n",
       "      <th>diffLength</th>\n",
       "      <th>TSS</th>\n",
       "      <th>cen_start</th>\n",
       "      <th>cen_end</th>\n",
       "      <th>TSS_in_arm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>49529869</td>\n",
       "      <td>49529985</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000236737</td>\n",
       "      <td>GAGE12B</td>\n",
       "      <td>116</td>\n",
       "      <td>18</td>\n",
       "      <td>49529985</td>\n",
       "      <td>58100000</td>\n",
       "      <td>63800000</td>\n",
       "      <td>q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>186401046</td>\n",
       "      <td>186401180</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000262180</td>\n",
       "      <td>OCLM</td>\n",
       "      <td>134</td>\n",
       "      <td>18</td>\n",
       "      <td>186401180</td>\n",
       "      <td>121700000</td>\n",
       "      <td>125100000</td>\n",
       "      <td>p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0 contig strand      start        end         biotype  \\\n",
       "0           0      X      +   49529869   49529985  protein_coding   \n",
       "1           1      1      +  186401046  186401180  protein_coding   \n",
       "\n",
       "           gene_id gene_name  length  diffLength        TSS  cen_start  \\\n",
       "0  ENSG00000236737   GAGE12B     116          18   49529985   58100000   \n",
       "1  ENSG00000262180      OCLM     134          18  186401180  121700000   \n",
       "\n",
       "     cen_end TSS_in_arm  \n",
       "0   63800000          q  \n",
       "1  125100000          p  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "p      13093\n",
       "q       6777\n",
       "cen       64\n",
       "Name: TSS_in_arm, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df.drop(genes_df.index[~genes_df[\"contig\"].isin(contig_l)], inplace=True)\n",
    "genes_df.shape\n",
    "\n",
    "genes_df = pd.merge(genes_df, cent_df, on=\"contig\", how=\"left\")\n",
    "\n",
    "genes_df[\"TSS_in_arm\"] = \"cen\"    \n",
    "genes_df.loc[genes_df['TSS'] < genes_df['cen_start'], 'TSS_in_arm'] = \"q\"\n",
    "genes_df.loc[genes_df['TSS'] > genes_df['cen_end']  , 'TSS_in_arm']   = \"p\"\n",
    "display(genes_df.head(2))\n",
    "genes_df[\"TSS_in_arm\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "92b77e34-9b19-44f8-8752-874ab0467e36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>42</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>6643</td>\n",
       "      <td>6450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>3385</td>\n",
       "      <td>3392</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand         +     -\n",
       "TSS_in_arm            \n",
       "cen           42    22\n",
       "p           6643  6450\n",
       "q           3385  3392"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(genes_df.TSS_in_arm, genes_df.strand)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a9b5cf28-8edf-4955-ac9e-bfcd65e41ecb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>481</td>\n",
       "      <td>496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>552</td>\n",
       "      <td>521</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "p           481  496\n",
       "q           552  521"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'10'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>306</td>\n",
       "      <td>268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>75</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           3    0\n",
       "p           306  268\n",
       "q            75   79"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'11'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>450</td>\n",
       "      <td>400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>211</td>\n",
       "      <td>235</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen          12    2\n",
       "p           450  400\n",
       "q           211  235"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'12'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>375</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>138</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           1    1\n",
       "p           375  369\n",
       "q           138  150"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'13'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>156</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "p           156  165"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'14'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>319</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "p           319  293"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'15'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>298</td>\n",
       "      <td>299</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "p           298  299"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'16'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>200</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>255</td>\n",
       "      <td>216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "p           200  181\n",
       "q           255  216"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'17'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>378</td>\n",
       "      <td>458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>158</td>\n",
       "      <td>189</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           1    0\n",
       "p           378  458\n",
       "q           158  189"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'18'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>100</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>40</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           0    1\n",
       "p           100  101\n",
       "q            40   26"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'19'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>424</td>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>310</td>\n",
       "      <td>330</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "p           424  407\n",
       "q           310  330"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'2'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>418</td>\n",
       "      <td>346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>232</td>\n",
       "      <td>238</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           9    4\n",
       "p           418  346\n",
       "q           232  238"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'20'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>198</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>92</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "p           198  162\n",
       "q            92   94"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'21'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>98</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand       +    -\n",
       "TSS_in_arm         \n",
       "p           98  118\n",
       "q            6   10"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'22'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>219</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           3    6\n",
       "p           219  215"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'3'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>281</td>\n",
       "      <td>289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>248</td>\n",
       "      <td>250</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           5    2\n",
       "p           281  289\n",
       "q           248  250"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'4'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>252</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>123</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           5    2\n",
       "p           252  269\n",
       "q           123  101"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'5'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>407</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>73</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           2    1\n",
       "p           407  319\n",
       "q            73   83"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'6'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>211</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>307</td>\n",
       "      <td>307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           0    1\n",
       "p           211  221\n",
       "q           307  307"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'7'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>328</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>142</td>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "p           328  293\n",
       "q           142  154"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'8'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>182</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>140</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           1    0\n",
       "p           182  228\n",
       "q           140  132"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'9'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>281</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>86</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "p           281  282\n",
       "q            86  129"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'X'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cen</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>263</td>\n",
       "      <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>184</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand        +    -\n",
       "TSS_in_arm          \n",
       "cen           0    2\n",
       "p           263  256\n",
       "q           184  146"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'Y'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strand</th>\n",
       "      <th>+</th>\n",
       "      <th>-</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSS_in_arm</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "strand       +   -\n",
       "TSS_in_arm        \n",
       "p           18  15\n",
       "q           13   2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for ch in sorted(contig_l):\n",
    "    aux_df = genes_df[genes_df.contig == ch]\n",
    "    display(ch)\n",
    "    display(pd.crosstab(aux_df.TSS_in_arm, aux_df.strand))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "722b896a-b4be-4981-89e9-c282064a3320",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
