{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "819bd65b-f5b5-4644-a058-0a8c843d4fe8",
   "metadata": {},
   "source": [
    "# Add a new column (of division) to the statistical descriptions\n",
    "in the protein coding gene length stat descriptions of the distributions for the different species. The column will be based in ensembl (division_7) and uniprot provided taxonomy (division_4_byLineage)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8046a185-31f3-48b1-83f7-947911a96d15",
   "metadata": {},
   "source": [
    "## Import python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "51b0a90c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os\n",
    "import sys\n",
    "\n",
    "\n",
    "gitHubProject_base_path = \"../../../\"\n",
    "main_tables_path  = gitHubProject_base_path + \"main_tables/\"\n",
    "if not os.path.isdir(main_tables_path):\n",
    "    print(\"The path (main_tables_path) is not defined! Check.\")  \n",
    "    print(main_tables_path, os.path.isdir(main_tables_path))\n",
    "    sys.exit(0)\n",
    "#    \n",
    "working_on_tables_path  = gitHubProject_base_path + \"working_on_tables/noisy/\"\n",
    "if not os.path.isdir(working_on_tables_path):\n",
    "    print(\"The path (working_on_tables_path) is not defined! Check.\")  \n",
    "    print(working_on_tables_path, os.path.isdir(working_on_tables_path))\n",
    "    sys.exit(0)\n",
    "#\n",
    "# Initialization:\n",
    "original_merged_file_name   = \"stat_merged.tsv\"\n",
    "noisy_merged_7671_file_name = \"stat_merged_7671.tsv\"\n",
    "\n",
    "BOOL_SAVE_FILE = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d509543a-9bbf-4428-bb98-a287201ef79d",
   "metadata": {},
   "source": [
    "## Retrieving the statistical descriptions \n",
    "**of the length's distributions for the different merged species**\n",
    "\n",
    "**Original:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e96ab72e-e702-4020-814a-625c86dd44ee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The statistical descriptions of the length distributions for the different species (merged set) is in:\n",
      " ../../../main_tables/stat_merged.tsv \n",
      "\n",
      "(6519, 66)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['genes_species', 'genes_assembly', 'genes_trunk_genes_path',\n",
       "       'genes_genes_file', 'genes_count', 'genes_mean', 'genes_std',\n",
       "       'genes_var', 'genes_min', 'genes_25perc', 'genes_50perc',\n",
       "       'genes_75perc', 'genes_max', 'genes_log10_mean', 'genes_log10_std',\n",
       "       'genes_log10_var', 'genes_log10_min', 'genes_log10_25perc',\n",
       "       'genes_log10_50perc', 'genes_log10_75perc', 'genes_log10_max',\n",
       "       'genes_log_mean', 'genes_log_std', 'genes_log_var', 'genes_log_min',\n",
       "       'genes_log_25perc', 'genes_log_50perc', 'genes_log_75perc',\n",
       "       'genes_log_max', 'genes_division', 'tax_id', 'prots_species',\n",
       "       'prots_proteome_id', 'prots_superregnum', 'prots_num_prot_cod_genes',\n",
       "       'prots_uniprot_fasta_file', 'prots_count', 'prots_mean', 'prots_std',\n",
       "       'prots_var', 'prots_min', 'prots_25perc', 'prots_50perc',\n",
       "       'prots_75perc', 'prots_max', 'prots_log10_mean', 'prots_log10_std',\n",
       "       'prots_log10_var', 'prots_log10_min', 'prots_log10_25perc',\n",
       "       'prots_log10_50perc', 'prots_log10_75perc', 'prots_log10_max',\n",
       "       'prots_log_mean', 'prots_log_std', 'prots_log_var', 'prots_log_min',\n",
       "       'prots_log_25perc', 'prots_log_50perc', 'prots_log_75perc',\n",
       "       'prots_log_max', 'merged_division_superregnum',\n",
       "       'ratio_mean_prots_genes', 'diff_prots_genes', 'abs_diff',\n",
       "       'ratio_prots_genes'],\n",
       "      dtype='object')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "merged_division_superregnum\n",
      "bacteria       5468\n",
      "fungi           533\n",
      "archaea         227\n",
      "vertebrates      92\n",
      "protist          91\n",
      "plants           59\n",
      "metazoa          49\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes_species</th>\n",
       "      <th>genes_assembly</th>\n",
       "      <th>genes_trunk_genes_path</th>\n",
       "      <th>genes_genes_file</th>\n",
       "      <th>genes_count</th>\n",
       "      <th>genes_mean</th>\n",
       "      <th>genes_std</th>\n",
       "      <th>genes_var</th>\n",
       "      <th>genes_min</th>\n",
       "      <th>genes_25perc</th>\n",
       "      <th>genes_50perc</th>\n",
       "      <th>genes_75perc</th>\n",
       "      <th>genes_max</th>\n",
       "      <th>genes_log10_mean</th>\n",
       "      <th>genes_log10_std</th>\n",
       "      <th>genes_log10_var</th>\n",
       "      <th>genes_log10_min</th>\n",
       "      <th>genes_log10_25perc</th>\n",
       "      <th>genes_log10_50perc</th>\n",
       "      <th>genes_log10_75perc</th>\n",
       "      <th>genes_log10_max</th>\n",
       "      <th>genes_log_mean</th>\n",
       "      <th>genes_log_std</th>\n",
       "      <th>genes_log_var</th>\n",
       "      <th>genes_log_min</th>\n",
       "      <th>genes_log_25perc</th>\n",
       "      <th>genes_log_50perc</th>\n",
       "      <th>genes_log_75perc</th>\n",
       "      <th>genes_log_max</th>\n",
       "      <th>genes_division</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>prots_species</th>\n",
       "      <th>prots_proteome_id</th>\n",
       "      <th>prots_superregnum</th>\n",
       "      <th>prots_num_prot_cod_genes</th>\n",
       "      <th>prots_uniprot_fasta_file</th>\n",
       "      <th>prots_count</th>\n",
       "      <th>prots_mean</th>\n",
       "      <th>prots_std</th>\n",
       "      <th>prots_var</th>\n",
       "      <th>prots_min</th>\n",
       "      <th>prots_25perc</th>\n",
       "      <th>prots_50perc</th>\n",
       "      <th>prots_75perc</th>\n",
       "      <th>prots_max</th>\n",
       "      <th>prots_log10_mean</th>\n",
       "      <th>prots_log10_std</th>\n",
       "      <th>prots_log10_var</th>\n",
       "      <th>prots_log10_min</th>\n",
       "      <th>prots_log10_25perc</th>\n",
       "      <th>prots_log10_50perc</th>\n",
       "      <th>prots_log10_75perc</th>\n",
       "      <th>prots_log10_max</th>\n",
       "      <th>prots_log_mean</th>\n",
       "      <th>prots_log_std</th>\n",
       "      <th>prots_log_var</th>\n",
       "      <th>prots_log_min</th>\n",
       "      <th>prots_log_25perc</th>\n",
       "      <th>prots_log_50perc</th>\n",
       "      <th>prots_log_75perc</th>\n",
       "      <th>prots_log_max</th>\n",
       "      <th>merged_division_superregnum</th>\n",
       "      <th>ratio_mean_prots_genes</th>\n",
       "      <th>diff_prots_genes</th>\n",
       "      <th>abs_diff</th>\n",
       "      <th>ratio_prots_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>archangium_gephyra_gca_001027285</td>\n",
       "      <td>ASM102728v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.archangium_gephyra_gca_00...</td>\n",
       "      <td>10119.0</td>\n",
       "      <td>1103.871035</td>\n",
       "      <td>1127.066255</td>\n",
       "      <td>1.270278e+06</td>\n",
       "      <td>113.0</td>\n",
       "      <td>524.0</td>\n",
       "      <td>887.0</td>\n",
       "      <td>1388.0</td>\n",
       "      <td>36653.0</td>\n",
       "      <td>2.919254</td>\n",
       "      <td>0.333869</td>\n",
       "      <td>0.111468</td>\n",
       "      <td>2.053078</td>\n",
       "      <td>2.719331</td>\n",
       "      <td>2.947924</td>\n",
       "      <td>3.142389</td>\n",
       "      <td>4.56411</td>\n",
       "      <td>6.721831</td>\n",
       "      <td>0.768762</td>\n",
       "      <td>0.590995</td>\n",
       "      <td>4.727388</td>\n",
       "      <td>6.261492</td>\n",
       "      <td>6.787845</td>\n",
       "      <td>7.235619</td>\n",
       "      <td>10.509251</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>48</td>\n",
       "      <td>Archangium gephyra</td>\n",
       "      <td>UP000035579</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>10110</td>\n",
       "      <td>/ftp.uniprot.org/pub/databases/uniprot/current...</td>\n",
       "      <td>10110.0</td>\n",
       "      <td>367.499703</td>\n",
       "      <td>375.772244</td>\n",
       "      <td>141204.779058</td>\n",
       "      <td>37.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>295.0</td>\n",
       "      <td>462.0</td>\n",
       "      <td>12217.0</td>\n",
       "      <td>2.441094</td>\n",
       "      <td>0.334836</td>\n",
       "      <td>0.112115</td>\n",
       "      <td>1.568202</td>\n",
       "      <td>2.240549</td>\n",
       "      <td>2.469822</td>\n",
       "      <td>2.664642</td>\n",
       "      <td>4.086965</td>\n",
       "      <td>5.620827</td>\n",
       "      <td>0.770987</td>\n",
       "      <td>0.594422</td>\n",
       "      <td>3.610918</td>\n",
       "      <td>5.159055</td>\n",
       "      <td>5.686975</td>\n",
       "      <td>6.135565</td>\n",
       "      <td>9.410584</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>0.998757</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.999111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chondromyces_crocatus_gca_001189295</td>\n",
       "      <td>ASM118929v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.chondromyces_crocatus_gca...</td>\n",
       "      <td>8339.0</td>\n",
       "      <td>1211.514450</td>\n",
       "      <td>1051.408605</td>\n",
       "      <td>1.105460e+06</td>\n",
       "      <td>89.0</td>\n",
       "      <td>629.0</td>\n",
       "      <td>977.0</td>\n",
       "      <td>1460.0</td>\n",
       "      <td>24476.0</td>\n",
       "      <td>2.978963</td>\n",
       "      <td>0.300250</td>\n",
       "      <td>0.090150</td>\n",
       "      <td>1.949390</td>\n",
       "      <td>2.798651</td>\n",
       "      <td>2.989895</td>\n",
       "      <td>3.164353</td>\n",
       "      <td>4.38874</td>\n",
       "      <td>6.859315</td>\n",
       "      <td>0.691352</td>\n",
       "      <td>0.477968</td>\n",
       "      <td>4.488636</td>\n",
       "      <td>6.444131</td>\n",
       "      <td>6.884487</td>\n",
       "      <td>7.286192</td>\n",
       "      <td>10.105448</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>52</td>\n",
       "      <td>Chondromyces crocatus</td>\n",
       "      <td>UP000067626</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>8327</td>\n",
       "      <td>/ftp.uniprot.org/pub/databases/uniprot/current...</td>\n",
       "      <td>8327.0</td>\n",
       "      <td>403.466555</td>\n",
       "      <td>350.579560</td>\n",
       "      <td>122906.027917</td>\n",
       "      <td>29.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>486.0</td>\n",
       "      <td>8158.0</td>\n",
       "      <td>2.501095</td>\n",
       "      <td>0.300984</td>\n",
       "      <td>0.090591</td>\n",
       "      <td>1.462398</td>\n",
       "      <td>2.320146</td>\n",
       "      <td>2.513218</td>\n",
       "      <td>2.686636</td>\n",
       "      <td>3.911584</td>\n",
       "      <td>5.758984</td>\n",
       "      <td>0.693041</td>\n",
       "      <td>0.480306</td>\n",
       "      <td>3.367296</td>\n",
       "      <td>5.342334</td>\n",
       "      <td>5.786897</td>\n",
       "      <td>6.186209</td>\n",
       "      <td>9.006754</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>0.999080</td>\n",
       "      <td>-12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.998561</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         genes_species genes_assembly  \\\n",
       "0     archangium_gephyra_gca_001027285    ASM102728v1   \n",
       "1  chondromyces_crocatus_gca_001189295    ASM118929v1   \n",
       "\n",
       "                              genes_trunk_genes_path  \\\n",
       "0  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "1  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "\n",
       "                                    genes_genes_file  genes_count  \\\n",
       "0  protein_coding.genes.archangium_gephyra_gca_00...      10119.0   \n",
       "1  protein_coding.genes.chondromyces_crocatus_gca...       8339.0   \n",
       "\n",
       "    genes_mean    genes_std     genes_var  genes_min  genes_25perc  \\\n",
       "0  1103.871035  1127.066255  1.270278e+06      113.0         524.0   \n",
       "1  1211.514450  1051.408605  1.105460e+06       89.0         629.0   \n",
       "\n",
       "   genes_50perc  genes_75perc  genes_max  genes_log10_mean  genes_log10_std  \\\n",
       "0         887.0        1388.0    36653.0          2.919254         0.333869   \n",
       "1         977.0        1460.0    24476.0          2.978963         0.300250   \n",
       "\n",
       "   genes_log10_var  genes_log10_min  genes_log10_25perc  genes_log10_50perc  \\\n",
       "0         0.111468         2.053078            2.719331            2.947924   \n",
       "1         0.090150         1.949390            2.798651            2.989895   \n",
       "\n",
       "   genes_log10_75perc  genes_log10_max  genes_log_mean  genes_log_std  \\\n",
       "0            3.142389          4.56411        6.721831       0.768762   \n",
       "1            3.164353          4.38874        6.859315       0.691352   \n",
       "\n",
       "   genes_log_var  genes_log_min  genes_log_25perc  genes_log_50perc  \\\n",
       "0       0.590995       4.727388          6.261492          6.787845   \n",
       "1       0.477968       4.488636          6.444131          6.884487   \n",
       "\n",
       "   genes_log_75perc  genes_log_max genes_division  tax_id  \\\n",
       "0          7.235619      10.509251       bacteria      48   \n",
       "1          7.286192      10.105448       bacteria      52   \n",
       "\n",
       "           prots_species prots_proteome_id prots_superregnum  \\\n",
       "0     Archangium gephyra       UP000035579          bacteria   \n",
       "1  Chondromyces crocatus       UP000067626          bacteria   \n",
       "\n",
       "   prots_num_prot_cod_genes  \\\n",
       "0                     10110   \n",
       "1                      8327   \n",
       "\n",
       "                            prots_uniprot_fasta_file  prots_count  prots_mean  \\\n",
       "0  /ftp.uniprot.org/pub/databases/uniprot/current...      10110.0  367.499703   \n",
       "1  /ftp.uniprot.org/pub/databases/uniprot/current...       8327.0  403.466555   \n",
       "\n",
       "    prots_std      prots_var  prots_min  prots_25perc  prots_50perc  \\\n",
       "0  375.772244  141204.779058       37.0         174.0         295.0   \n",
       "1  350.579560  122906.027917       29.0         209.0         326.0   \n",
       "\n",
       "   prots_75perc  prots_max  prots_log10_mean  prots_log10_std  \\\n",
       "0         462.0    12217.0          2.441094         0.334836   \n",
       "1         486.0     8158.0          2.501095         0.300984   \n",
       "\n",
       "   prots_log10_var  prots_log10_min  prots_log10_25perc  prots_log10_50perc  \\\n",
       "0         0.112115         1.568202            2.240549            2.469822   \n",
       "1         0.090591         1.462398            2.320146            2.513218   \n",
       "\n",
       "   prots_log10_75perc  prots_log10_max  prots_log_mean  prots_log_std  \\\n",
       "0            2.664642         4.086965        5.620827       0.770987   \n",
       "1            2.686636         3.911584        5.758984       0.693041   \n",
       "\n",
       "   prots_log_var  prots_log_min  prots_log_25perc  prots_log_50perc  \\\n",
       "0       0.594422       3.610918          5.159055          5.686975   \n",
       "1       0.480306       3.367296          5.342334          5.786897   \n",
       "\n",
       "   prots_log_75perc  prots_log_max merged_division_superregnum  \\\n",
       "0          6.135565       9.410584                    bacteria   \n",
       "1          6.186209       9.006754                    bacteria   \n",
       "\n",
       "   ratio_mean_prots_genes  diff_prots_genes  abs_diff  ratio_prots_genes  \n",
       "0                0.998757              -9.0       9.0           0.999111  \n",
       "1                0.999080             -12.0      12.0           0.998561  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../../working_on_tables/ True\n",
      "to be saved in (not yet):  ../../../working_on_tables/noisy_preprocessed_stat_merged.tsv\n",
      "saving...\n",
      "...file saved. Done!\n"
     ]
    }
   ],
   "source": [
    "# statistics on length distribution for different species\n",
    "merged_file = main_tables_path + original_merged_file_name\n",
    "print(\"The statistical descriptions of the length distributions for the different species (merged set) is in:\\n\", merged_file, \"\\n\")\n",
    "\n",
    "# retrieve data and diminish the number of columns\n",
    "merged_df = pd.read_csv(merged_file, low_memory=False, sep=\"\\t\")\n",
    "merged_taxid_l = merged_df[\"tax_id\"].to_list()\n",
    "\n",
    "if 1: # visualize the data frame\n",
    "    print(merged_df.shape)\n",
    "    display(merged_df.columns)\n",
    "    print(merged_df.value_counts('merged_division_superregnum', dropna=False))\n",
    "    pd.set_option('display.max_columns', merged_df.shape[1])\n",
    "    display(merged_df.head(2))\n",
    "\n",
    "\n",
    "# eliminate the columns that the noisy species (7671) do not have\n",
    "merged_df = merged_df.drop(columns=['ratio_mean_prots_genes', 'diff_prots_genes', 'abs_diff', 'ratio_prots_genes'])\n",
    "#\n",
    "print(working_on_tables_path, os.path.isdir(working_on_tables_path))\n",
    "out_original_file = working_on_tables_path + \"noisy_preprocessed_\" + original_merged_file_name\n",
    "print(\"to be saved in (not yet): \", out_original_file)\n",
    "if BOOL_SAVE_FILE: # Save the file if desired\n",
    "    print(\"saving...\")\n",
    "    merged_df.to_csv(out_original_file, \n",
    "                     sep=\"\\t\", index=False, header=True)\n",
    "    print(\"...file saved. Done!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d1b21d9-ac06-4be4-959e-680642a3459c",
   "metadata": {},
   "source": [
    "## Get noisy species\n",
    "and eliminate the ones that are not noisy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "92ee192e-1890-4460-be9d-ef1306d52887",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The statistical descriptions of the length distributions for the different species (merged set) is in:\n",
      " ../../../working_on_tables/stat_merged_7671.tsv \n",
      "\n",
      "(1152, 62)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['genes_species', 'genes_assembly', 'genes_trunk_genes_path',\n",
       "       'genes_genes_file', 'genes_count', 'genes_mean', 'genes_std',\n",
       "       'genes_var', 'genes_min', 'genes_25perc', 'genes_50perc',\n",
       "       'genes_75perc', 'genes_max', 'genes_log10_mean', 'genes_log10_std',\n",
       "       'genes_log10_var', 'genes_log10_min', 'genes_log10_25perc',\n",
       "       'genes_log10_50perc', 'genes_log10_75perc', 'genes_log10_max',\n",
       "       'genes_log_mean', 'genes_log_std', 'genes_log_var', 'genes_log_min',\n",
       "       'genes_log_25perc', 'genes_log_50perc', 'genes_log_75perc',\n",
       "       'genes_log_max', 'genes_division', 'tax_id', 'prots_species',\n",
       "       'prots_proteome_id', 'prots_superregnum', 'prots_num_prot_cod_genes',\n",
       "       'prots_uniprot_fasta_file', 'prots_count', 'prots_mean', 'prots_std',\n",
       "       'prots_var', 'prots_min', 'prots_25perc', 'prots_50perc',\n",
       "       'prots_75perc', 'prots_max', 'prots_log10_mean', 'prots_log10_std',\n",
       "       'prots_log10_var', 'prots_log10_min', 'prots_log10_25perc',\n",
       "       'prots_log10_50perc', 'prots_log10_75perc', 'prots_log10_max',\n",
       "       'prots_log_mean', 'prots_log_std', 'prots_log_var', 'prots_log_min',\n",
       "       'prots_log_25perc', 'prots_log_50perc', 'prots_log_75perc',\n",
       "       'prots_log_max', 'merged_division_superregnum'],\n",
       "      dtype='object')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "merged_division_superregnum\n",
      "bacteria       991\n",
      "archaea         56\n",
      "fungi           33\n",
      "protist         23\n",
      "vertebrates     22\n",
      "metazoa         14\n",
      "plants          13\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes_species</th>\n",
       "      <th>genes_assembly</th>\n",
       "      <th>genes_trunk_genes_path</th>\n",
       "      <th>genes_genes_file</th>\n",
       "      <th>genes_count</th>\n",
       "      <th>genes_mean</th>\n",
       "      <th>genes_std</th>\n",
       "      <th>genes_var</th>\n",
       "      <th>genes_min</th>\n",
       "      <th>genes_25perc</th>\n",
       "      <th>genes_50perc</th>\n",
       "      <th>genes_75perc</th>\n",
       "      <th>genes_max</th>\n",
       "      <th>genes_log10_mean</th>\n",
       "      <th>genes_log10_std</th>\n",
       "      <th>genes_log10_var</th>\n",
       "      <th>genes_log10_min</th>\n",
       "      <th>genes_log10_25perc</th>\n",
       "      <th>genes_log10_50perc</th>\n",
       "      <th>genes_log10_75perc</th>\n",
       "      <th>genes_log10_max</th>\n",
       "      <th>genes_log_mean</th>\n",
       "      <th>genes_log_std</th>\n",
       "      <th>genes_log_var</th>\n",
       "      <th>genes_log_min</th>\n",
       "      <th>genes_log_25perc</th>\n",
       "      <th>genes_log_50perc</th>\n",
       "      <th>genes_log_75perc</th>\n",
       "      <th>genes_log_max</th>\n",
       "      <th>genes_division</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>prots_species</th>\n",
       "      <th>prots_proteome_id</th>\n",
       "      <th>prots_superregnum</th>\n",
       "      <th>prots_num_prot_cod_genes</th>\n",
       "      <th>prots_uniprot_fasta_file</th>\n",
       "      <th>prots_count</th>\n",
       "      <th>prots_mean</th>\n",
       "      <th>prots_std</th>\n",
       "      <th>prots_var</th>\n",
       "      <th>prots_min</th>\n",
       "      <th>prots_25perc</th>\n",
       "      <th>prots_50perc</th>\n",
       "      <th>prots_75perc</th>\n",
       "      <th>prots_max</th>\n",
       "      <th>prots_log10_mean</th>\n",
       "      <th>prots_log10_std</th>\n",
       "      <th>prots_log10_var</th>\n",
       "      <th>prots_log10_min</th>\n",
       "      <th>prots_log10_25perc</th>\n",
       "      <th>prots_log10_50perc</th>\n",
       "      <th>prots_log10_75perc</th>\n",
       "      <th>prots_log10_max</th>\n",
       "      <th>prots_log_mean</th>\n",
       "      <th>prots_log_std</th>\n",
       "      <th>prots_log_var</th>\n",
       "      <th>prots_log_min</th>\n",
       "      <th>prots_log_25perc</th>\n",
       "      <th>prots_log_50perc</th>\n",
       "      <th>prots_log_75perc</th>\n",
       "      <th>prots_log_max</th>\n",
       "      <th>merged_division_superregnum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>vitreoscilla_filiformis_gca_002222655</td>\n",
       "      <td>ASM222265v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.vitreoscilla_filiformis_g...</td>\n",
       "      <td>3557.0</td>\n",
       "      <td>976.152376</td>\n",
       "      <td>716.082928</td>\n",
       "      <td>512774.759677</td>\n",
       "      <td>122.0</td>\n",
       "      <td>488.0</td>\n",
       "      <td>818.0</td>\n",
       "      <td>1241.0</td>\n",
       "      <td>7925.0</td>\n",
       "      <td>2.887739</td>\n",
       "      <td>0.305841</td>\n",
       "      <td>0.093539</td>\n",
       "      <td>2.08636</td>\n",
       "      <td>2.688420</td>\n",
       "      <td>2.912753</td>\n",
       "      <td>3.093772</td>\n",
       "      <td>3.898999</td>\n",
       "      <td>6.649266</td>\n",
       "      <td>0.704226</td>\n",
       "      <td>0.495934</td>\n",
       "      <td>4.804021</td>\n",
       "      <td>6.190315</td>\n",
       "      <td>6.706862</td>\n",
       "      <td>7.123673</td>\n",
       "      <td>8.977778</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>63</td>\n",
       "      <td>Vitreoscilla filiformis</td>\n",
       "      <td>UP000199729</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>3396</td>\n",
       "      <td>/ftp.uniprot.org/pub/databases/uniprot/current...</td>\n",
       "      <td>3396.0</td>\n",
       "      <td>327.099529</td>\n",
       "      <td>239.207207</td>\n",
       "      <td>57220.087882</td>\n",
       "      <td>40.0</td>\n",
       "      <td>166.0</td>\n",
       "      <td>275.5</td>\n",
       "      <td>414.0</td>\n",
       "      <td>2641.0</td>\n",
       "      <td>2.413182</td>\n",
       "      <td>0.306324</td>\n",
       "      <td>0.093835</td>\n",
       "      <td>1.602060</td>\n",
       "      <td>2.220108</td>\n",
       "      <td>2.440121</td>\n",
       "      <td>2.617000</td>\n",
       "      <td>3.421768</td>\n",
       "      <td>5.556556</td>\n",
       "      <td>0.705337</td>\n",
       "      <td>0.497501</td>\n",
       "      <td>3.688879</td>\n",
       "      <td>5.111988</td>\n",
       "      <td>5.618586</td>\n",
       "      <td>6.025866</td>\n",
       "      <td>7.878913</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gemmata_obscuriglobus_gca_003149495</td>\n",
       "      <td>ASM314949v1</td>\n",
       "      <td>ftp.ensemblgenomes.org/pub/bacteria/release-49...</td>\n",
       "      <td>protein_coding.genes.gemmata_obscuriglobus_gca...</td>\n",
       "      <td>6994.0</td>\n",
       "      <td>1032.628253</td>\n",
       "      <td>851.139209</td>\n",
       "      <td>724437.953590</td>\n",
       "      <td>80.0</td>\n",
       "      <td>458.0</td>\n",
       "      <td>845.0</td>\n",
       "      <td>1313.0</td>\n",
       "      <td>19658.0</td>\n",
       "      <td>2.900650</td>\n",
       "      <td>0.313935</td>\n",
       "      <td>0.098555</td>\n",
       "      <td>1.90309</td>\n",
       "      <td>2.660865</td>\n",
       "      <td>2.926857</td>\n",
       "      <td>3.118265</td>\n",
       "      <td>4.293539</td>\n",
       "      <td>6.678994</td>\n",
       "      <td>0.722863</td>\n",
       "      <td>0.522531</td>\n",
       "      <td>4.382027</td>\n",
       "      <td>6.126869</td>\n",
       "      <td>6.739337</td>\n",
       "      <td>7.180070</td>\n",
       "      <td>9.886240</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>114</td>\n",
       "      <td>Gemmata obscuriglobus</td>\n",
       "      <td>UP000245802</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>6809</td>\n",
       "      <td>/ftp.uniprot.org/pub/databases/uniprot/current...</td>\n",
       "      <td>6809.0</td>\n",
       "      <td>345.550007</td>\n",
       "      <td>286.373330</td>\n",
       "      <td>82009.684081</td>\n",
       "      <td>26.0</td>\n",
       "      <td>154.0</td>\n",
       "      <td>282.0</td>\n",
       "      <td>438.0</td>\n",
       "      <td>6552.0</td>\n",
       "      <td>2.423783</td>\n",
       "      <td>0.316089</td>\n",
       "      <td>0.099912</td>\n",
       "      <td>1.414973</td>\n",
       "      <td>2.187521</td>\n",
       "      <td>2.450249</td>\n",
       "      <td>2.641474</td>\n",
       "      <td>3.816374</td>\n",
       "      <td>5.580967</td>\n",
       "      <td>0.727822</td>\n",
       "      <td>0.529726</td>\n",
       "      <td>3.258097</td>\n",
       "      <td>5.036953</td>\n",
       "      <td>5.641907</td>\n",
       "      <td>6.082219</td>\n",
       "      <td>8.787526</td>\n",
       "      <td>bacteria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           genes_species genes_assembly  \\\n",
       "3  vitreoscilla_filiformis_gca_002222655    ASM222265v1   \n",
       "7    gemmata_obscuriglobus_gca_003149495    ASM314949v1   \n",
       "\n",
       "                              genes_trunk_genes_path  \\\n",
       "3  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "7  ftp.ensemblgenomes.org/pub/bacteria/release-49...   \n",
       "\n",
       "                                    genes_genes_file  genes_count  \\\n",
       "3  protein_coding.genes.vitreoscilla_filiformis_g...       3557.0   \n",
       "7  protein_coding.genes.gemmata_obscuriglobus_gca...       6994.0   \n",
       "\n",
       "    genes_mean   genes_std      genes_var  genes_min  genes_25perc  \\\n",
       "3   976.152376  716.082928  512774.759677      122.0         488.0   \n",
       "7  1032.628253  851.139209  724437.953590       80.0         458.0   \n",
       "\n",
       "   genes_50perc  genes_75perc  genes_max  genes_log10_mean  genes_log10_std  \\\n",
       "3         818.0        1241.0     7925.0          2.887739         0.305841   \n",
       "7         845.0        1313.0    19658.0          2.900650         0.313935   \n",
       "\n",
       "   genes_log10_var  genes_log10_min  genes_log10_25perc  genes_log10_50perc  \\\n",
       "3         0.093539          2.08636            2.688420            2.912753   \n",
       "7         0.098555          1.90309            2.660865            2.926857   \n",
       "\n",
       "   genes_log10_75perc  genes_log10_max  genes_log_mean  genes_log_std  \\\n",
       "3            3.093772         3.898999        6.649266       0.704226   \n",
       "7            3.118265         4.293539        6.678994       0.722863   \n",
       "\n",
       "   genes_log_var  genes_log_min  genes_log_25perc  genes_log_50perc  \\\n",
       "3       0.495934       4.804021          6.190315          6.706862   \n",
       "7       0.522531       4.382027          6.126869          6.739337   \n",
       "\n",
       "   genes_log_75perc  genes_log_max genes_division  tax_id  \\\n",
       "3          7.123673       8.977778       bacteria      63   \n",
       "7          7.180070       9.886240       bacteria     114   \n",
       "\n",
       "             prots_species prots_proteome_id prots_superregnum  \\\n",
       "3  Vitreoscilla filiformis       UP000199729          bacteria   \n",
       "7    Gemmata obscuriglobus       UP000245802          bacteria   \n",
       "\n",
       "   prots_num_prot_cod_genes  \\\n",
       "3                      3396   \n",
       "7                      6809   \n",
       "\n",
       "                            prots_uniprot_fasta_file  prots_count  prots_mean  \\\n",
       "3  /ftp.uniprot.org/pub/databases/uniprot/current...       3396.0  327.099529   \n",
       "7  /ftp.uniprot.org/pub/databases/uniprot/current...       6809.0  345.550007   \n",
       "\n",
       "    prots_std     prots_var  prots_min  prots_25perc  prots_50perc  \\\n",
       "3  239.207207  57220.087882       40.0         166.0         275.5   \n",
       "7  286.373330  82009.684081       26.0         154.0         282.0   \n",
       "\n",
       "   prots_75perc  prots_max  prots_log10_mean  prots_log10_std  \\\n",
       "3         414.0     2641.0          2.413182         0.306324   \n",
       "7         438.0     6552.0          2.423783         0.316089   \n",
       "\n",
       "   prots_log10_var  prots_log10_min  prots_log10_25perc  prots_log10_50perc  \\\n",
       "3         0.093835         1.602060            2.220108            2.440121   \n",
       "7         0.099912         1.414973            2.187521            2.450249   \n",
       "\n",
       "   prots_log10_75perc  prots_log10_max  prots_log_mean  prots_log_std  \\\n",
       "3            2.617000         3.421768        5.556556       0.705337   \n",
       "7            2.641474         3.816374        5.580967       0.727822   \n",
       "\n",
       "   prots_log_var  prots_log_min  prots_log_25perc  prots_log_50perc  \\\n",
       "3       0.497501       3.688879          5.111988          5.618586   \n",
       "7       0.529726       3.258097          5.036953          5.641907   \n",
       "\n",
       "   prots_log_75perc  prots_log_max merged_division_superregnum  \n",
       "3          6.025866       7.878913                    bacteria  \n",
       "7          6.082219       8.787526                    bacteria  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# statistics on length distribution for different species\n",
    "noisy_7671_file = working_on_tables_path + noisy_merged_7671_file_name\n",
    "print(\"The statistical descriptions of the length distributions for the different species (merged set) is in:\\n\", noisy_7671_file, \"\\n\")\n",
    "\n",
    "# retrieve data and diminish the number of columns\n",
    "noisy_df = pd.read_csv(noisy_7671_file, low_memory=False, sep=\"\\t\")\n",
    "noisy_df = noisy_df.drop(\n",
    "    noisy_df[noisy_df[\"tax_id\"].isin(merged_taxid_l)].index\n",
    ")\n",
    "if 1: # visualize the data frame\n",
    "    print(noisy_df.shape)\n",
    "    display(noisy_df.columns)\n",
    "    print(noisy_df.value_counts('merged_division_superregnum', dropna=False))\n",
    "    pd.set_option('display.max_columns', noisy_df.shape[1])\n",
    "    display(noisy_df.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91edd84d-6592-4dad-9b06-6553a28fcbf4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5a3b0bd7-4c5b-4285-8a0e-4d23d0298cb7",
   "metadata": {},
   "source": [
    "### Save the file\n",
    "**if desired** (be careful not to overwrite any file!).\n",
    "Use **BOOL_SAVE_FILE variable** above for that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aa8f75b4-cdea-4d26-83e4-e8c017ce4dd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../../working_on_tables/ True\n",
      "to be saved in (not yet):  ../../../working_on_tables/diff_6519_and_stat_merged_7671.tsv\n",
      "saving...\n",
      "...file saved. Done!\n"
     ]
    }
   ],
   "source": [
    "print(working_on_tables_path, os.path.isdir(working_on_tables_path))\n",
    "out_file = working_on_tables_path + \"diff_6519_and_\" + noisy_merged_7671_file_name\n",
    "print(\"to be saved in (not yet): \", out_file)\n",
    "if BOOL_SAVE_FILE: # Save the file if desired\n",
    "    print(\"saving...\")\n",
    "    noisy_df.to_csv(out_file, \n",
    "                   sep=\"\\t\", index=False, header=False)\n",
    "    print(\"...file saved. Done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10a93a33-1243-4cd8-ad40-a460322ad633",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
